{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import secrets\n",
    "import pandas as pd\n",
    "import scipy.stats as spstats\n",
    "import statsmodels.api as sm\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm import tqdm\n",
    "from IPython.display import display\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_dir(dir: str):\n",
    "  \"\"\"\n",
    "  Creates a given path directory \"dir\" if it does not exist.\n",
    "  Args:\n",
    "    dir (str): Path to the directory.\n",
    "  \"\"\"\n",
    "  if os.path.exists(dir) and os.path.isdir(dir):\n",
    "    pass\n",
    "  else:\n",
    "    os.makedirs(dir)\n",
    "\n",
    "\n",
    "def spearman_corr(expression, mutation_burden):\n",
    "  rho, pval = spstats.spearmanr(\n",
    "    expression,\n",
    "    mutation_burden,\n",
    "  )\n",
    "  return pd.Series({\"rho\": rho, \"pval\": pval,})\n",
    "\n",
    "\n",
    "def driver_neighbour_corr(\n",
    "  driver: int,\n",
    "  neighbourlist: np.ndarray,\n",
    "  mutationtab: np.ndarray,\n",
    "  expressiontab: np.ndarray,\n",
    "  ctfilter: np.ndarray | None = None,\n",
    ") -> tuple:\n",
    "  \"\"\"\n",
    "  Computes spearman rank correlation coefficient.\n",
    "\n",
    "  Parameters:\n",
    "  -----------\n",
    "  driver: driver index to index in mutationtab\n",
    "  \n",
    "  neighbourlist: 1D (listlike) with neighbour indices to index expressiontab\n",
    "  \n",
    "  mutationtab: array with driver mutations of shape\n",
    "      (# cancer types, # drivers)\n",
    "      \n",
    "  expressiontab: array with neighbour expression of shape\n",
    "      (# cancer types, # neighbours)\n",
    "      \n",
    "  ctfilter: boolean array to remove cancer types without mutation in mutationtab\n",
    "  and expressiontab with shape (# cancer types, # drivers)\n",
    "\n",
    "  Returns:\n",
    "  --------\n",
    "  tuple with shape (rho list, p-value list, neighbour list)\n",
    "  \"\"\"\n",
    "    \n",
    "  if ctfilter is not None:\n",
    "    filt = ctfilter[driver]\n",
    "      \n",
    "  else:\n",
    "    filt = [True] * len(mutationtab)\n",
    "\n",
    "  exp = expressiontab[np.ix_(filt, neighbourlist)]\n",
    "\n",
    "  # remove neighbours without expression after cancer type filter\n",
    "  mask = np.flatnonzero(exp.sum(axis=0) > 0)\n",
    "    \n",
    "  if len(mask) == 0:\n",
    "    # if driver does not have neighbours with expression\n",
    "    return [np.nan], [np.nan], [np.nan]\n",
    "      \n",
    "  elif len(mask) == 1:\n",
    "    # if driver has only one neighbour with expression\n",
    "    rho, pvalue = spstats.spearmanr(exp[:, mask], mutationtab[filt, driver])\n",
    "    return [rho], [pvalue], neighbourlist[mask]\n",
    "  \n",
    "  else:\n",
    "    rho, pvalue = spstats.spearmanr(exp[:, mask], mutationtab[filt, driver])\n",
    "    return rho[-1, :-1], pvalue[-1, :-1], neighbourlist[mask]\n",
    "\n",
    "\n",
    "def between_cancer_corr(\n",
    "  exparray: np.ndarray,\n",
    "  mutationarray: np.ndarray,\n",
    "  graph: pd.DataFrame,\n",
    "  cancertype_filter: np.ndarray | None = None,\n",
    "  n_jobs: int = -1,\n",
    "  progressbar: bool = True,\n",
    "):\n",
    "  \"\"\"\n",
    "  Computes Between Cancer Type Analysis\n",
    "  \n",
    "  Parameters:\n",
    "  -----------\n",
    "  exparray: array with neighbour expression of shape\n",
    "      (# cancer types, # neighbours)\n",
    "  \n",
    "  graph: boolean array with shape (# neighbours, # drivers)\n",
    "  \n",
    "  mutationarray: array with driver mutations of shape\n",
    "      (# cancer types, # drivers)\n",
    "      \n",
    "  cancertype_filter: boolean array to remove cancer types without mutation in\n",
    "      mutationtab and expressiontab with shape (# cancer types, # drivers)\n",
    "\n",
    "  Returns:\n",
    "  --------\n",
    "  pd.DataFrame with columns [driver, neighbour, rho, rho_pval]\n",
    "  \"\"\"\n",
    "    \n",
    "  driverlist = graph.columns.tolist()\n",
    "  if progressbar:\n",
    "    neighbourlist = tqdm([np.flatnonzero(graph[driver]) for driver in driverlist])\n",
    "      \n",
    "  else:\n",
    "    neighbourlist = [np.flatnonzero(graph[driver]) for driver in driverlist]\n",
    "\n",
    "  # compute correlation\n",
    "  corr_results = Parallel(n_jobs=n_jobs)(\n",
    "    delayed(driver_neighbour_corr)\n",
    "    (driver, neighbours, mutationarray, exparray, cancertype_filter)\n",
    "    for driver, neighbours in enumerate(neighbourlist)\n",
    "  )\n",
    "\n",
    "  results = {\"driver\": [], \"neighbour\": [], \"rho\": [], \"rho_pval\": [],}\n",
    "  for driver in range(len(driverlist)):\n",
    "    results[\"driver\"].extend([driverlist[driver]] * len(corr_results[driver][2]))\n",
    "    results[\"neighbour\"].extend(corr_results[driver][2])\n",
    "    results[\"rho\"].extend(corr_results[driver][0])\n",
    "    results[\"rho_pval\"].extend(corr_results[driver][1])\n",
    "\n",
    "  results = (\n",
    "    pd.DataFrame(results)\n",
    "    .set_index(\"neighbour\")\n",
    "    .merge(\n",
    "      pd.Series(graph.index).rename(\"neighbour\"),\n",
    "      left_index=True,\n",
    "      right_index=True,\n",
    "    )\n",
    "    .dropna()\n",
    "    .reset_index(drop=True)\n",
    "  )\n",
    "  return results\n",
    "\n",
    "\n",
    "def ols(y, x, index_x=False):\n",
    "  \"\"\"\n",
    "  Ordinary Least Squares regression.\n",
    "  Args:\n",
    "      y (np.ndarray): Dependent variable.\n",
    "      x (np.ndarray): Independent variable.\n",
    "      index_x (bool): Set True if x is a DataFrame with the same index as y.\n",
    "  Returns:\n",
    "      tuple: OLS coefficients and residuals.\n",
    "  \"\"\"\n",
    "  if index_x:\n",
    "    x = x[y.name]\n",
    "  filt = y>0\n",
    "  x = sm.add_constant(x)\n",
    "  model = sm.OLS(y[filt], x[filt]).fit()\n",
    "  return model.resid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rawdata = \"data/raw/\"\n",
    "outputdata = \"data/processed/\"\n",
    "shuffled_dir = outputdata+\"shuffled_bct/\"\n",
    "check_dir(shuffled_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Mutation Burden"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2625385, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>effect</th>\n",
       "      <th>patient</th>\n",
       "      <th>nonsilent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>Silent</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene             effect       patient  nonsilent\n",
       "0    TACC2  Missense_Mutation  TCGA-02-0003       True\n",
       "1  JAKMIP3             Silent  TCGA-02-0003      False\n",
       "2    PANX3  Missense_Mutation  TCGA-02-0003       True\n",
       "3     SPI1  Missense_Mutation  TCGA-02-0003       True\n",
       "4  NAALAD2  Missense_Mutation  TCGA-02-0003       True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silent_mutation = [\"Silent\", \"5'Flank\", \"3'Flank\", \"5'UTR\", \"3'UTR\", \"Intron\", \"RNA\"]\n",
    "mutationtab = (\n",
    "    pd.read_table(rawdata+\"mc3.v0.2.8.PUBLIC.xena\",)\n",
    "    [[\"sample\", \"gene\", \"effect\"]]\n",
    "    .assign(\n",
    "      patient=lambda x: x[\"sample\"].str.split(\"-\").str[:-1].str.join(\"-\"),\n",
    "      nonsilent=lambda x: ~x[\"effect\"].isin(silent_mutation))\n",
    "    .drop(columns=[\"sample\"])\n",
    "    .drop_duplicates()\n",
    ")\n",
    "print(mutationtab.shape)\n",
    "mutationtab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# patients: 7317\n",
      "# cancer types: 32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047</th>\n",
       "      <th>GBM</th>\n",
       "      <td>6.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.05</td>\n",
       "      <td>5.40</td>\n",
       "      <td>5.22</td>\n",
       "      <td>1.16</td>\n",
       "      <td>8.87</td>\n",
       "      <td>8.92</td>\n",
       "      <td>7.87</td>\n",
       "      <td>10.01</td>\n",
       "      <td>...</td>\n",
       "      <td>8.03</td>\n",
       "      <td>8.66</td>\n",
       "      <td>6.05</td>\n",
       "      <td>8.48</td>\n",
       "      <td>10.12</td>\n",
       "      <td>0.69</td>\n",
       "      <td>10.24</td>\n",
       "      <td>11.92</td>\n",
       "      <td>10.45</td>\n",
       "      <td>9.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055</th>\n",
       "      <th>GBM</th>\n",
       "      <td>8.62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.39</td>\n",
       "      <td>1.42</td>\n",
       "      <td>8.93</td>\n",
       "      <td>0.64</td>\n",
       "      <td>9.22</td>\n",
       "      <td>8.31</td>\n",
       "      <td>6.66</td>\n",
       "      <td>10.41</td>\n",
       "      <td>...</td>\n",
       "      <td>8.87</td>\n",
       "      <td>7.95</td>\n",
       "      <td>5.45</td>\n",
       "      <td>8.14</td>\n",
       "      <td>9.25</td>\n",
       "      <td>2.60</td>\n",
       "      <td>9.85</td>\n",
       "      <td>13.49</td>\n",
       "      <td>9.25</td>\n",
       "      <td>9.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483</th>\n",
       "      <th>GBM</th>\n",
       "      <td>8.09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.36</td>\n",
       "      <td>1.82</td>\n",
       "      <td>6.46</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.11</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.02</td>\n",
       "      <td>9.92</td>\n",
       "      <td>...</td>\n",
       "      <td>9.42</td>\n",
       "      <td>9.39</td>\n",
       "      <td>4.35</td>\n",
       "      <td>8.67</td>\n",
       "      <td>9.76</td>\n",
       "      <td>5.50</td>\n",
       "      <td>10.24</td>\n",
       "      <td>12.31</td>\n",
       "      <td>9.70</td>\n",
       "      <td>9.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485</th>\n",
       "      <th>GBM</th>\n",
       "      <td>6.41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.93</td>\n",
       "      <td>7.73</td>\n",
       "      <td>7.29</td>\n",
       "      <td>0.56</td>\n",
       "      <td>9.99</td>\n",
       "      <td>8.25</td>\n",
       "      <td>7.58</td>\n",
       "      <td>10.36</td>\n",
       "      <td>...</td>\n",
       "      <td>8.79</td>\n",
       "      <td>8.79</td>\n",
       "      <td>5.78</td>\n",
       "      <td>8.10</td>\n",
       "      <td>10.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.06</td>\n",
       "      <td>12.31</td>\n",
       "      <td>10.16</td>\n",
       "      <td>9.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486</th>\n",
       "      <th>GBM</th>\n",
       "      <td>6.77</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.32</td>\n",
       "      <td>6.71</td>\n",
       "      <td>5.49</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.46</td>\n",
       "      <td>8.62</td>\n",
       "      <td>7.77</td>\n",
       "      <td>10.54</td>\n",
       "      <td>...</td>\n",
       "      <td>7.39</td>\n",
       "      <td>6.24</td>\n",
       "      <td>5.03</td>\n",
       "      <td>7.64</td>\n",
       "      <td>9.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.43</td>\n",
       "      <td>12.93</td>\n",
       "      <td>9.30</td>\n",
       "      <td>9.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          A1BG  A1CF    A2M  A2ML1  A4GALT  A4GNT   AAAS  \\\n",
       "patient      cancer_type                                                   \n",
       "TCGA-02-0047 GBM          6.98   0.0  15.05   5.40    5.22   1.16   8.87   \n",
       "TCGA-02-0055 GBM          8.62   0.0  15.39   1.42    8.93   0.64   9.22   \n",
       "TCGA-02-2483 GBM          8.09   0.0  14.36   1.82    6.46   0.00  10.11   \n",
       "TCGA-02-2485 GBM          6.41   0.0  12.93   7.73    7.29   0.56   9.99   \n",
       "TCGA-02-2486 GBM          6.77   0.0  15.32   6.71    5.49   0.00   9.46   \n",
       "\n",
       "                          AACS  AADAT  AAGAB  ...  ZWILCH  ZWINT  ZXDA  ZXDB  \\\n",
       "patient      cancer_type                      ...                              \n",
       "TCGA-02-0047 GBM          8.92   7.87  10.01  ...    8.03   8.66  6.05  8.48   \n",
       "TCGA-02-0055 GBM          8.31   6.66  10.41  ...    8.87   7.95  5.45  8.14   \n",
       "TCGA-02-2483 GBM          8.95   8.02   9.92  ...    9.42   9.39  4.35  8.67   \n",
       "TCGA-02-2485 GBM          8.25   7.58  10.36  ...    8.79   8.79  5.78  8.10   \n",
       "TCGA-02-2486 GBM          8.62   7.77  10.54  ...    7.39   6.24  5.03  7.64   \n",
       "\n",
       "                           ZXDC  ZYG11A  ZYG11B    ZYX  ZZEF1  ZZZ3  \n",
       "patient      cancer_type                                             \n",
       "TCGA-02-0047 GBM          10.12    0.69   10.24  11.92  10.45  9.24  \n",
       "TCGA-02-0055 GBM           9.25    2.60    9.85  13.49   9.25  9.49  \n",
       "TCGA-02-2483 GBM           9.76    5.50   10.24  12.31   9.70  9.46  \n",
       "TCGA-02-2485 GBM          10.40    0.00   10.06  12.31  10.16  9.45  \n",
       "TCGA-02-2486 GBM           9.35    0.00    9.43  12.93   9.30  9.05  \n",
       "\n",
       "[5 rows x 15464 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-02-0047</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-02-0055</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-02-2483</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-02-2485</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-02-2486</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        patient cancer_type\n",
       "0  TCGA-02-0047         GBM\n",
       "1  TCGA-02-0055         GBM\n",
       "2  TCGA-02-2483         GBM\n",
       "3  TCGA-02-2485         GBM\n",
       "4  TCGA-02-2486         GBM"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exptab = pd.read_feather(outputdata+\"expression.feather\")\n",
    "clintab = exptab.index.to_frame(index=False)\n",
    "print(\"# patients:\", len(clintab))\n",
    "print(\"# cancer types:\", clintab.cancer_type.nunique())\n",
    "display(exptab.head())\n",
    "clintab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# drivers: 3081\n",
      "# neighbours: 15464\n"
     ]
    }
   ],
   "source": [
    "graph = pd.read_csv(outputdata+\"main_graph.csv\")\n",
    "drivers = graph[\"driver\"].unique()\n",
    "neighbours = graph[\"neighbour\"].unique()\n",
    "print(\"# drivers:\", len(drivers))\n",
    "print(\"# neighbours:\", len(neighbours))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Mutation Burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_burden = (\n",
    "    mutationtab\n",
    "    .groupby(\"patient\", as_index=False)\n",
    "    [\"nonsilent\"]\n",
    "    .sum()\n",
    "    .merge(clintab, on=\"patient\")\n",
    "    .rename(columns={\"nonsilent\": \"mutation_burden\"})\n",
    "    [[\"patient\", \"cancer_type\", \"mutation_burden\"]]\n",
    ")\n",
    "print(mutation_burden.shape)\n",
    "mutation_burden.to_csv(outputdata+\"mutation_burden.csv\", index=False)\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Driver-specific mutation burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_burden = (\n",
    "    mutationtab\n",
    "    .merge(clintab, on=\"patient\")\n",
    "    .groupby([\"cancer_type\", \"gene\"], as_index=False)\n",
    "    [\"nonsilent\"]\n",
    "    # mean mutation burden per pair driver-cancer type\n",
    "    .mean()\n",
    "    .pivot(index=\"cancer_type\", columns=\"gene\", values=\"nonsilent\")\n",
    "    [drivers]\n",
    "    .fillna(0)\n",
    "    .reset_index()\n",
    ")\n",
    "print(mutation_burden.shape)\n",
    "mutation_burden.to_csv(outputdata+\"driver_specific_mutation_burden.csv\", index=False)\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Correlation Between Mutation Burden and Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_burden = pd.read_csv(outputdata+\"mutation_burden.csv\")\n",
    "print(mutation_burden.shape)\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corrtab, mutation_burden = (\n",
    "  exptab\n",
    "  .align(mutation_burden.set_index([\"patient\", \"cancer_type\"]), join=\"inner\", axis=0, level=0)\n",
    " )\n",
    "print(corrtab.shape)\n",
    "print(mutation_burden.shape)\n",
    "display(corrtab.head())\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation = (\n",
    "  corrtab\n",
    "    .apply(\n",
    "      lambda x: spearman_corr(\n",
    "        x.to_numpy().flatten(),\n",
    "      mutation_burden[\"mutation_burden\"].to_numpy()\n",
    "      ),\n",
    "      axis=0,\n",
    "    )\n",
    "  .T\n",
    ")\n",
    "print(correlation.shape)\n",
    "print(len(correlation.dropna()))\n",
    "correlation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "signcorr = correlation[correlation[\"pval\"] < 0.05]\n",
    "print(len(signcorr))\n",
    "print(\"(+) correlation\", len(signcorr[signcorr[\"rho\"] > 0]))\n",
    "print(\"(-) correlation\", len(signcorr[signcorr[\"rho\"] < 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation.to_csv(outputdata+\"mutation_burden_corr.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Between Cancer Types Associations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15464, 3081)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>driver</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>ABCA10</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCA7</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>ABCC3</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCC9</th>\n",
       "      <th>...</th>\n",
       "      <th>ZRANB3</th>\n",
       "      <th>ZRSR2</th>\n",
       "      <th>ZSCAN31</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSWIM3</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neighbour</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3081 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "driver      A1CF  A2ML1  ABCA10  ABCA13  ABCA7  ABCB1  ABCB5  ABCC3  ABCC5  \\\n",
       "neighbour                                                                    \n",
       "A1BG       False  False   False   False  False  False  False  False  False   \n",
       "A1CF       False  False   False   False  False  False  False  False  False   \n",
       "A2M        False  False   False   False  False  False  False  False  False   \n",
       "A2ML1      False  False   False   False  False  False  False  False  False   \n",
       "A4GALT     False  False   False   False  False  False  False  False  False   \n",
       "\n",
       "driver     ABCC9  ...  ZRANB3  ZRSR2  ZSCAN31  ZSCAN4  ZSWIM3  ZSWIM6  ZSWIM7  \\\n",
       "neighbour         ...                                                           \n",
       "A1BG       False  ...   False  False    False   False   False   False   False   \n",
       "A1CF       False  ...   False  False    False   False   False   False   False   \n",
       "A2M        False  ...   False  False    False   False   False   False   False   \n",
       "A2ML1      False  ...   False  False    False   False   False   False   False   \n",
       "A4GALT     False  ...   False  False    False   False   False   False   False   \n",
       "\n",
       "driver     ZWILCH  ZWINT  ZZEF1  \n",
       "neighbour                        \n",
       "A1BG        False  False  False  \n",
       "A1CF        False  False  False  \n",
       "A2M         False  False  False  \n",
       "A2ML1       False  False  False  \n",
       "A4GALT      False  False  False  \n",
       "\n",
       "[5 rows x 3081 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph = pd.read_feather(outputdata+\"neighbours.feather\")\n",
    "print(graph.shape)\n",
    "graph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACC</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CESC</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHOL</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cancer_type  freq\n",
       "0         ACC    73\n",
       "1        BLCA   288\n",
       "2        BRCA   760\n",
       "3        CESC   253\n",
       "4        CHOL    35"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 3081)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>ABCA10</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCA7</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>ABCC3</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCC9</th>\n",
       "      <th>...</th>\n",
       "      <th>ZRANB3</th>\n",
       "      <th>ZRSR2</th>\n",
       "      <th>ZSCAN31</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSWIM3</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CESC</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHOL</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3081 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             A1CF  A2ML1  ABCA10  ABCA13  ABCA7  ABCB1  ABCB5  ABCC3  ABCC5  \\\n",
       "cancer_type                                                                   \n",
       "ACC             0      0       0       1      0      0      1      1      0   \n",
       "BLCA            1      6       4      19      7      4      8      7      5   \n",
       "BRCA            1      5       6      15      7      4      5      4      6   \n",
       "CESC            3      4       1      10      3      1      3      3      2   \n",
       "CHOL            1      0       0       1      0      0      0      0      0   \n",
       "\n",
       "             ABCC9  ...  ZRANB3  ZRSR2  ZSCAN31  ZSCAN4  ZSWIM3  ZSWIM6  \\\n",
       "cancer_type         ...                                                   \n",
       "ACC              1  ...       2      1        0       0       0       1   \n",
       "BLCA            12  ...       5      2        0       0       0       0   \n",
       "BRCA            10  ...       2      0        0       0       4       2   \n",
       "CESC             5  ...       2      1        0       1       0       0   \n",
       "CHOL             0  ...       0      0        0       0       0       0   \n",
       "\n",
       "             ZSWIM7  ZWILCH  ZWINT  ZZEF1  \n",
       "cancer_type                                \n",
       "ACC               0       0      0      0  \n",
       "BLCA              1       2      0     10  \n",
       "BRCA              0       3      2      4  \n",
       "CESC              0       2      1      2  \n",
       "CHOL              0       0      0      0  \n",
       "\n",
       "[5 rows x 3081 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutationtab = pd.read_feather(outputdata+\"mutation.feather\")\n",
    "# number of individuals with mutations per cancer type\n",
    "cancer_freq = (\n",
    "  mutationtab.index.to_frame(index=False)\n",
    "  .groupby(\"cancer_type\", as_index=False)\n",
    "  .size()\n",
    "  .rename(columns={\"size\": \"freq\"})\n",
    ")\n",
    "# number of mutated individuals per cancer type\n",
    "mutationtab = (\n",
    "  mutationtab\n",
    "  .groupby([\"cancer_type\"])\n",
    "  .sum()\n",
    ")\n",
    "\n",
    "display(cancer_freq.head())\n",
    "print(mutationtab.shape)\n",
    "mutationtab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 15464)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>5.851781</td>\n",
       "      <td>0.135616</td>\n",
       "      <td>13.047671</td>\n",
       "      <td>3.620548</td>\n",
       "      <td>7.893425</td>\n",
       "      <td>0.310274</td>\n",
       "      <td>10.721507</td>\n",
       "      <td>10.201781</td>\n",
       "      <td>5.040000</td>\n",
       "      <td>10.077123</td>\n",
       "      <td>...</td>\n",
       "      <td>7.492740</td>\n",
       "      <td>8.361507</td>\n",
       "      <td>4.685342</td>\n",
       "      <td>7.888904</td>\n",
       "      <td>9.919589</td>\n",
       "      <td>1.240137</td>\n",
       "      <td>9.573425</td>\n",
       "      <td>10.610822</td>\n",
       "      <td>9.883151</td>\n",
       "      <td>8.889315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>5.422882</td>\n",
       "      <td>0.433264</td>\n",
       "      <td>12.652847</td>\n",
       "      <td>7.206562</td>\n",
       "      <td>9.460729</td>\n",
       "      <td>0.675278</td>\n",
       "      <td>9.714722</td>\n",
       "      <td>9.418819</td>\n",
       "      <td>7.122535</td>\n",
       "      <td>10.239965</td>\n",
       "      <td>...</td>\n",
       "      <td>8.883403</td>\n",
       "      <td>9.703507</td>\n",
       "      <td>4.867361</td>\n",
       "      <td>8.038889</td>\n",
       "      <td>9.965208</td>\n",
       "      <td>3.223333</td>\n",
       "      <td>9.590139</td>\n",
       "      <td>12.098681</td>\n",
       "      <td>9.826944</td>\n",
       "      <td>9.302361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>7.128882</td>\n",
       "      <td>0.109618</td>\n",
       "      <td>13.495566</td>\n",
       "      <td>3.458474</td>\n",
       "      <td>8.081000</td>\n",
       "      <td>0.671026</td>\n",
       "      <td>9.417474</td>\n",
       "      <td>9.965684</td>\n",
       "      <td>5.593118</td>\n",
       "      <td>10.629211</td>\n",
       "      <td>...</td>\n",
       "      <td>8.774500</td>\n",
       "      <td>9.377132</td>\n",
       "      <td>5.894934</td>\n",
       "      <td>8.994039</td>\n",
       "      <td>10.075250</td>\n",
       "      <td>6.206553</td>\n",
       "      <td>9.729579</td>\n",
       "      <td>11.800447</td>\n",
       "      <td>10.150566</td>\n",
       "      <td>9.807053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CESC</th>\n",
       "      <td>5.663557</td>\n",
       "      <td>0.382530</td>\n",
       "      <td>11.309130</td>\n",
       "      <td>8.673913</td>\n",
       "      <td>9.790198</td>\n",
       "      <td>0.755810</td>\n",
       "      <td>9.707510</td>\n",
       "      <td>10.105059</td>\n",
       "      <td>5.265534</td>\n",
       "      <td>10.303043</td>\n",
       "      <td>...</td>\n",
       "      <td>9.702885</td>\n",
       "      <td>10.714466</td>\n",
       "      <td>5.028024</td>\n",
       "      <td>8.102213</td>\n",
       "      <td>10.281028</td>\n",
       "      <td>6.270435</td>\n",
       "      <td>9.254150</td>\n",
       "      <td>11.797036</td>\n",
       "      <td>9.832688</td>\n",
       "      <td>9.449921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHOL</th>\n",
       "      <td>7.678286</td>\n",
       "      <td>7.446000</td>\n",
       "      <td>12.957429</td>\n",
       "      <td>0.789429</td>\n",
       "      <td>8.179714</td>\n",
       "      <td>3.774571</td>\n",
       "      <td>9.802286</td>\n",
       "      <td>9.980000</td>\n",
       "      <td>5.859714</td>\n",
       "      <td>10.088000</td>\n",
       "      <td>...</td>\n",
       "      <td>8.234857</td>\n",
       "      <td>8.757429</td>\n",
       "      <td>5.628286</td>\n",
       "      <td>8.532286</td>\n",
       "      <td>10.021429</td>\n",
       "      <td>1.162857</td>\n",
       "      <td>9.445429</td>\n",
       "      <td>11.936571</td>\n",
       "      <td>10.002857</td>\n",
       "      <td>9.471714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 A1BG      A1CF        A2M     A2ML1    A4GALT     A4GNT  \\\n",
       "cancer_type                                                                \n",
       "ACC          5.851781  0.135616  13.047671  3.620548  7.893425  0.310274   \n",
       "BLCA         5.422882  0.433264  12.652847  7.206562  9.460729  0.675278   \n",
       "BRCA         7.128882  0.109618  13.495566  3.458474  8.081000  0.671026   \n",
       "CESC         5.663557  0.382530  11.309130  8.673913  9.790198  0.755810   \n",
       "CHOL         7.678286  7.446000  12.957429  0.789429  8.179714  3.774571   \n",
       "\n",
       "                  AAAS       AACS     AADAT      AAGAB  ...    ZWILCH  \\\n",
       "cancer_type                                             ...             \n",
       "ACC          10.721507  10.201781  5.040000  10.077123  ...  7.492740   \n",
       "BLCA          9.714722   9.418819  7.122535  10.239965  ...  8.883403   \n",
       "BRCA          9.417474   9.965684  5.593118  10.629211  ...  8.774500   \n",
       "CESC          9.707510  10.105059  5.265534  10.303043  ...  9.702885   \n",
       "CHOL          9.802286   9.980000  5.859714  10.088000  ...  8.234857   \n",
       "\n",
       "                 ZWINT      ZXDA      ZXDB       ZXDC    ZYG11A    ZYG11B  \\\n",
       "cancer_type                                                                 \n",
       "ACC           8.361507  4.685342  7.888904   9.919589  1.240137  9.573425   \n",
       "BLCA          9.703507  4.867361  8.038889   9.965208  3.223333  9.590139   \n",
       "BRCA          9.377132  5.894934  8.994039  10.075250  6.206553  9.729579   \n",
       "CESC         10.714466  5.028024  8.102213  10.281028  6.270435  9.254150   \n",
       "CHOL          8.757429  5.628286  8.532286  10.021429  1.162857  9.445429   \n",
       "\n",
       "                   ZYX      ZZEF1      ZZZ3  \n",
       "cancer_type                                  \n",
       "ACC          10.610822   9.883151  8.889315  \n",
       "BLCA         12.098681   9.826944  9.302361  \n",
       "BRCA         11.800447  10.150566  9.807053  \n",
       "CESC         11.797036   9.832688  9.449921  \n",
       "CHOL         11.936571  10.002857  9.471714  \n",
       "\n",
       "[5 rows x 15464 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exptab = (\n",
    "    pd.read_feather(outputdata+\"expression.feather\")\n",
    "    .groupby([\"cancer_type\"])\n",
    "    .mean()\n",
    ")\n",
    "print(exptab.shape)\n",
    "exptab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_burden</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>34.712329</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>119.857639</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>46.130263</td>\n",
       "      <td>760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CESC</th>\n",
       "      <td>87.233202</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHOL</th>\n",
       "      <td>33.828571</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation_burden  freq\n",
       "cancer_type                       \n",
       "ACC                34.712329    73\n",
       "BLCA              119.857639   288\n",
       "BRCA               46.130263   760\n",
       "CESC               87.233202   253\n",
       "CHOL               33.828571    35"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_burden = (\n",
    "  pd.read_csv(outputdata+\"mutation_burden.csv\")\n",
    "  .groupby(\"cancer_type\", as_index=False)\n",
    "  [\"mutation_burden\"]\n",
    "  .mean()\n",
    "  .merge(cancer_freq, on=\"cancer_type\")\n",
    "  .set_index(\"cancer_type\")\n",
    "  .sort_index()\n",
    "  .rename(columns=lambda x: x.lower())\n",
    ")\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 3081)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>ABCA10</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCA7</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>ABCC3</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCC9</th>\n",
       "      <th>...</th>\n",
       "      <th>ZRANB3</th>\n",
       "      <th>ZRSR2</th>\n",
       "      <th>ZSCAN31</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSWIM3</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CESC</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>...</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHOL</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3081 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 A1CF     A2ML1    ABCA10    ABCA13     ABCA7     ABCB1  \\\n",
       "cancer_type                                                               \n",
       "ACC          0.000000  0.000000  0.000000  0.500000  0.000000  0.000000   \n",
       "BLCA         0.333333  0.545455  0.666667  0.833333  0.727273  0.800000   \n",
       "BRCA         1.000000  0.555556  1.000000  0.625000  1.000000  0.500000   \n",
       "CESC         1.000000  0.666667  0.333333  0.625000  0.428571  0.333333   \n",
       "CHOL         1.000000  0.000000  0.000000  1.000000  0.000000  0.000000   \n",
       "\n",
       "                ABCB5     ABCC3     ABCC5     ABCC9  ...    ZRANB3  ZRSR2  \\\n",
       "cancer_type                                          ...                    \n",
       "ACC          1.000000  0.500000  0.000000  0.500000  ...  1.000000    1.0   \n",
       "BLCA         0.888889  0.777778  0.625000  0.750000  ...  1.000000    1.0   \n",
       "BRCA         0.714286  0.571429  0.750000  0.769231  ...  0.500000    0.0   \n",
       "CESC         1.000000  0.428571  0.222222  0.454545  ...  0.666667    0.5   \n",
       "CHOL         0.000000  0.000000  0.000000  0.000000  ...  0.000000    0.0   \n",
       "\n",
       "             ZSCAN31  ZSCAN4  ZSWIM3    ZSWIM6  ZSWIM7    ZWILCH  ZWINT  \\\n",
       "cancer_type                                                               \n",
       "ACC              0.0     0.0     0.0  1.000000     0.0  0.000000    0.0   \n",
       "BLCA             0.0     0.0     0.0  0.000000     1.0  1.000000    0.0   \n",
       "BRCA             0.0     0.0     0.8  0.666667     0.0  1.000000    1.0   \n",
       "CESC             0.0     0.5     0.0  0.000000     0.0  0.666667    1.0   \n",
       "CHOL             0.0     0.0     0.0  0.000000     0.0  0.000000    0.0   \n",
       "\n",
       "                ZZEF1  \n",
       "cancer_type            \n",
       "ACC          0.000000  \n",
       "BLCA         0.714286  \n",
       "BRCA         1.000000  \n",
       "CESC         0.666667  \n",
       "CHOL         0.000000  \n",
       "\n",
       "[5 rows x 3081 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specific_mutation_burden = pd.read_csv(\n",
    "  outputdata+\"driver_specific_mutation_burden.csv\", index_col=\"cancer_type\")\n",
    "print(specific_mutation_burden.shape)\n",
    "specific_mutation_burden.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exptab: (32, 15464)\n",
      "mutationtab: (32, 3081)\n",
      "mutation_burden: (32, 2)\n",
      "specific_mutation_burden: (32, 3081)\n",
      "graph: (15464, 3081)\n"
     ]
    }
   ],
   "source": [
    "print(\"exptab:\", exptab.shape)\n",
    "print(\"mutationtab:\", mutationtab.shape)\n",
    "print(\"mutation_burden:\", mutation_burden.shape)\n",
    "print(\"specific_mutation_burden:\", specific_mutation_burden.shape)\n",
    "print(\"graph:\", graph.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Spearman Correlation with real pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[False False False ... False False False]\n",
      " [ True  True  True ...  True False  True]\n",
      " [ True  True  True ...  True  True  True]\n",
      " ...\n",
      " [False  True  True ... False  True  True]\n",
      " [False  True False ... False False False]\n",
      " [False False False ... False False False]]\n"
     ]
    }
   ],
   "source": [
    "exparray = (\n",
    "    exptab\n",
    "    .sort_index(level=0)\n",
    "    .to_numpy()\n",
    ")\n",
    "mutationarray = (\n",
    "    mutationtab\n",
    "    # compute relative frequency\n",
    "    .div(mutation_burden.freq, axis=0)\n",
    "    # order axis\n",
    "    .T.sort_index()\n",
    "    .T.sort_index(level=0)\n",
    ")\n",
    "# filter to remove cancer types with no mutations prior to correlation calculation\n",
    "cancertype_filter = (\n",
    "  mutationarray.T\n",
    "  .reset_index(names=\"driver\")\n",
    "  .set_index(\"driver\")\n",
    "  # remove cancer types with no mutations\n",
    "  .transform(lambda x: x > 0)\n",
    ")\n",
    "cancertype_filter.to_csv(outputdata+\"cancertype_filter.csv\", index=True)\n",
    "cancertype_filter = cancertype_filter.to_numpy()\n",
    "print(cancertype_filter.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bct_corr: (32, 3081)\n",
      "bct_corr_mb: (32, 3081)\n",
      "bct_corr_mb_specific: (32, 3081)\n",
      "bct_corr_mb_logspecific: (32, 3081)\n"
     ]
    }
   ],
   "source": [
    "normalization_dict = {\n",
    "  \"bct_corr\": mutationarray.to_numpy(),\n",
    "  \"bct_corr_mb\": mutationarray.div(mutation_burden.mutation_burden, axis=0).to_numpy(),\n",
    "  \"bct_corr_mb_specific\": (mutationarray/specific_mutation_burden).fillna(0).to_numpy(),\n",
    "  \"bct_corr_mb_logspecific\": (\n",
    "    (mutationarray/np.log10(specific_mutation_burden+1))\n",
    "    .fillna(0)\n",
    "    .to_numpy()\n",
    "  )\n",
    "}\n",
    "for key, value in normalization_dict.items():\n",
    "    print(f\"{key}: {value.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exparray: (32, 15464)\n",
      "mutationarray: (32, 3081)\n",
      "specific_mutation_burden: (32, 3081)\n",
      "graph: (15464, 3081)\n",
      "cancertype_filter: (3081, 32)\n"
     ]
    }
   ],
   "source": [
    "print(\"exparray:\", exparray.shape)\n",
    "print(\"mutationarray:\", mutationarray.shape)\n",
    "print(\"specific_mutation_burden:\", specific_mutation_burden.shape)\n",
    "print(\"graph:\", graph.shape)\n",
    "print(\"cancertype_filter:\", cancertype_filter.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nct</th>\n",
       "      <th>nmut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>21</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA10</th>\n",
       "      <td>20</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA13</th>\n",
       "      <td>28</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA7</th>\n",
       "      <td>20</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        nct  nmut\n",
       "A1CF     16    39\n",
       "A2ML1    21    67\n",
       "ABCA10   20    73\n",
       "ABCA13   28   266\n",
       "ABCA7    20    66"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmut = mutationtab.sum(axis=0).sort_index()\n",
    "nct = pd.Series(cancertype_filter.sum(axis=1), index=nmut.index)\n",
    "nct_nmut = pd.concat([nct, nmut], axis=1).rename(columns={0: \"nct\", 1: \"nmut\"})\n",
    "nct_nmut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Compute Correlation: all cancer types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for norm, mutarray in tqdm(normalization_dict.items()):\n",
    "    print(norm)\n",
    "    # compute correlation\n",
    "    corr = between_cancer_corr(\n",
    "        exparray,\n",
    "        mutarray,\n",
    "        graph,\n",
    "        cancertype_filter,\n",
    "    )\n",
    "    # save results\n",
    "    (\n",
    "      corr\n",
    "      .merge(nct_nmut, left_on=\"driver\", right_index=True)\n",
    "      .to_csv(outputdata+f\"{norm}.csv\", index=False)\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Compute Correlation: OLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutarray = (\n",
    "  mutationarray\n",
    "  .apply(ols, axis=0, args=(mutation_burden.mutation_burden,))\n",
    "  .fillna(0)\n",
    "  .to_numpy()\n",
    ")\n",
    "corr = between_cancer_corr(\n",
    "        exparray,\n",
    "        mutarray,\n",
    "        graph,\n",
    "        cancertype_filter,\n",
    "    )\n",
    "corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutarray = (\n",
    "  mutationarray\n",
    "  .apply(ols, axis=0, args=(specific_mutation_burden, True))\n",
    "  .fillna(0)\n",
    "  .to_numpy()\n",
    ")\n",
    "spec_corr = between_cancer_corr(\n",
    "        exparray,\n",
    "        mutarray,\n",
    "        graph,\n",
    "        cancertype_filter,\n",
    "    )\n",
    "spec_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtgraph = pd.read_csv(outputdata+\"main_graph_filtered.csv\")\n",
    "pd.merge(corr, filtgraph).to_csv(outputdata+\"bct_corr_mb_ols.csv\", index=False)\n",
    "pd.merge(spec_corr, filtgraph).to_csv(outputdata+\"bct_corr_mb_specific_ols.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Compute Correlation: low and high mutation frequency groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutarray = normalization_dict[\"bct_corr_mb_specific\"]\n",
    "median = np.nanmedian(np.where(mutarray>0, mutarray, np.nan), axis=0)\n",
    "low_mut_filter = cancertype_filter & (mutarray <= median).T\n",
    "high_mut_filter = cancertype_filter & (mutarray >= median).T\n",
    "driver_indexer = np.flatnonzero(high_mut_filter.sum(axis=1)>2)\n",
    "\n",
    "mutarray = mutarray[:, driver_indexer]\n",
    "low_mut_group = low_mut_filter[driver_indexer]\n",
    "high_mut_group = high_mut_filter[driver_indexer]\n",
    "filtgraph = graph.iloc[:, driver_indexer]\n",
    "print(\"mutarray:\", mutarray.shape)\n",
    "print(\"low_mut_group:\", low_mut_group.shape)\n",
    "print(\"high_mut_group:\", high_mut_group.shape)\n",
    "print(\"filtgraph:\", filtgraph.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_results = []\n",
    "for group, filter in zip([\"low_mut\", \"high_mut\"], [low_mut_group, high_mut_group]):\n",
    "    print(group)\n",
    "    # compute correlation\n",
    "    corr = between_cancer_corr(\n",
    "        exparray,\n",
    "        mutarray,\n",
    "        filtgraph,\n",
    "        filter,\n",
    "    )\n",
    "    # save results\n",
    "    corr_results.append(corr)\n",
    "\n",
    "corr_results = pd.merge(\n",
    "    *corr_results,\n",
    "    on=[\"driver\", \"neighbour\"],\n",
    "    suffixes=(\"_low_mut\", \"_high_mut\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_results.to_csv(outputdata+\"bct_corr_mutation_groups.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Spearman Correlation with fake pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fakepairs = (\n",
    "  pd.read_csv(outputdata+\"full_pathway_analysis.csv\", usecols=[\"driver\", \"fake_neighbour\"])\n",
    "  .dropna()\n",
    ")\n",
    "drivers = fakepairs.driver.unique()\n",
    "fake_neighbours = fakepairs.fake_neighbour.unique()\n",
    "print(\"# drivers:\", len(drivers))\n",
    "print(\"# fake neighbours:\", len(fake_neighbours))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exparray = (\n",
    "    exptab[fake_neighbours]\n",
    "    .sort_index(level=0)\n",
    "    .to_numpy()\n",
    ")\n",
    "mutationarray = (\n",
    "    mutationtab[drivers]\n",
    "    # compute relative frequency\n",
    "    .div(mutation_burden.freq, axis=0)\n",
    "    # order axis\n",
    "    .T.sort_index()\n",
    "    .T.sort_index(level=0)\n",
    ")\n",
    "# filter to remove cancer types with no mutations prior to correlation calculation\n",
    "cancertype_filter = (\n",
    "  mutationarray[drivers].T\n",
    "  .reset_index(names=\"driver\")\n",
    "  .set_index(\"driver\")\n",
    "  # remove cancer types with no mutations\n",
    "  .transform(lambda x: x > 0)\n",
    "  .to_numpy()\n",
    ")\n",
    "mutationarray = (mutationarray/specific_mutation_burden[drivers]).fillna(0).to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fake_graph = (\n",
    "  fakepairs\n",
    "  .assign(value=1)\n",
    "  .pivot_table(index=\"fake_neighbour\", columns=\"driver\", values=\"value\", fill_value=0)\n",
    "  .astype(bool)\n",
    ")\n",
    "fake_graph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"exparray:\", exparray.shape)\n",
    "print(\"mutationarray:\", mutationarray.shape)\n",
    "print(\"fake graph:\", fake_graph.shape)\n",
    "print(\"cancertype_filter:\", cancertype_filter.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Compute Correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute correlation\n",
    "corr = between_cancer_corr(\n",
    "  exparray,\n",
    "  mutationarray,\n",
    "  fake_graph,\n",
    "  cancertype_filter,\n",
    ")\n",
    "# save results\n",
    "corr.to_csv(outputdata+\"bct_corr_fake_pairs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Filtered BCT Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maingraph = pd.read_csv(outputdata+\"main_graph.csv\")\n",
    "pairs_to_exclude = pd.read_csv(outputdata+\"pairs_to_exclude_bct.csv\")\n",
    "print(\"# pairs in main graph:\", len(maingraph))\n",
    "print(\"# pairs to exclude:\", len(pairs_to_exclude))\n",
    "display(maingraph.head())\n",
    "pairs_to_exclude.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredpairs = (\n",
    "  maingraph\n",
    "  .merge(pairs_to_exclude, how=\"outer\", indicator=True)\n",
    "  .query(\"_merge == 'left_only'\")\n",
    "  .drop(columns=[\"_merge\"])\n",
    ")\n",
    "print(\"pairs in filtered graph:\", len(filteredpairs))\n",
    "print(\"# drivers:\", filteredpairs.driver.nunique())\n",
    "print(\"# neighbours:\", filteredpairs.neighbour.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bct_corr = pd.read_csv(outputdata+\"bct_corr_mb_specific.csv\")\n",
    "print(len(bct_corr))\n",
    "bct_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_bct_corr = (\n",
    "  bct_corr.merge(filteredpairs)\n",
    "  [[\"driver\", \"neighbour\", \"rho\", \"rho_pval\", \"nct\", \"nmut\"]]\n",
    ")\n",
    "print(\"# pairs:\", len(filtered_bct_corr))\n",
    "print(\"# significant pairs:\", len(filtered_bct_corr[filtered_bct_corr.rho_pval < 0.05]))\n",
    "print(\"# drivers:\", filtered_bct_corr.driver.nunique())\n",
    "print(\"# neighbours:\", filtered_bct_corr.neighbour.nunique())\n",
    "filtered_bct_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredpairs.to_csv(outputdata+\"main_graph_filtered.csv\", index=False)\n",
    "filtered_bct_corr.to_csv(outputdata+\"bct_corr_filtered.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Random shuffles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shufflepairs = pd.read_csv(outputdata+\"main_graph_filtered.csv\")\n",
    "drivers = shufflepairs.driver.unique()\n",
    "neighbours = shufflepairs.neighbour.unique()\n",
    "print(\"# shuffle pairs:\", len(shufflepairs))\n",
    "print(\"# drivers:\", len(drivers))\n",
    "print(\"# neighbours:\", len(neighbours))\n",
    "shufflepairs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exptab = (\n",
    "  pd.read_feather(outputdata+\"expression.feather\")\n",
    "  [neighbours]\n",
    "  # drop patient labels\n",
    "  .droplevel(0)\n",
    ")\n",
    "print(exptab.shape)\n",
    "exptab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutationtab = pd.read_feather(outputdata+\"mutation.feather\")\n",
    "\n",
    "cancer_freq = (\n",
    "  mutationtab.index.to_frame(index=False)\n",
    "  .groupby(\"cancer_type\")\n",
    "  .size()\n",
    ")\n",
    "\n",
    "mutationtab = (\n",
    "  mutationtab[drivers]\n",
    "  .groupby([\"cancer_type\"])\n",
    "  .sum()\n",
    "  .div(cancer_freq, axis=0)\n",
    "  # order axis\n",
    "  .T.sort_index()\n",
    "  .T.sort_index()\n",
    ")\n",
    "print(cancer_freq.shape)\n",
    "print(mutationtab.shape)\n",
    "mutationtab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_burden = (\n",
    "    pd.read_csv(outputdata+\"mutation_burden.csv\")\n",
    "    .groupby(\"cancer_type\", as_index=False)\n",
    "    [\"mutation_burden\"]\n",
    "    .mean()\n",
    "    .set_index(\"cancer_type\")\n",
    "    .sort_index()\n",
    "    .mutation_burden\n",
    ")\n",
    "specific_mutation_burden = (\n",
    "  pd.read_csv(\n",
    "    outputdata+\"driver_specific_mutation_burden.csv\", index_col=\"cancer_type\")\n",
    "  [drivers]\n",
    ")\n",
    "display(mutation_burden.head(2))\n",
    "specific_mutation_burden.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"exptab:\", exptab.shape)\n",
    "print(\"mutationtab:\", mutationtab.shape)\n",
    "print(\"mutation_burden:\", mutation_burden.shape)\n",
    "print(\"specific_mutation_burden:\", specific_mutation_burden.shape)\n",
    "print(\"graph:\", graph.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter to remove cancer types with no mutations prior to correlation calculation\n",
    "cancertype_filter = (\n",
    "  mutationtab.T\n",
    "  .reset_index(names=\"driver\")\n",
    "  .set_index(\"driver\")\n",
    "  # remove cancer types with no mutations\n",
    "  .transform(lambda x: x > 0)\n",
    "  .to_numpy()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffle_graph = (\n",
    "  shufflepairs\n",
    "  .assign(value=1)\n",
    "  .pivot_table(index=\"neighbour\", columns=\"driver\", values=\"value\", fill_value=0)\n",
    "  .astype(bool)\n",
    ")\n",
    "shuffle_graph.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## shuffle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutationarray = (mutationtab/specific_mutation_burden[drivers]).fillna(0).to_numpy()\n",
    "print(\"mutationarray:\", mutationarray.shape)\n",
    "print(\"shuffle graph:\", shuffle_graph.shape)\n",
    "print(\"cancertype_filter:\", cancertype_filter.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(secrets.randbits(128))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 61028791052787875347939602848649680021\n",
    "rng = np.random.default_rng(seed)\n",
    "spawn = rng.spawn(100)\n",
    "index = exptab.index.to_numpy()\n",
    "\n",
    "for i, child_rng in enumerate(tqdm(spawn)):\n",
    "  # shuffle expression data\n",
    "  shuffled_index = pd.Index(child_rng.permutation(index), name=\"cancer_type\")\n",
    "  exparray = (\n",
    "    exptab\n",
    "    .set_index(shuffled_index)\n",
    "    .groupby([\"cancer_type\"])\n",
    "    .mean()\n",
    "    .sort_index()\n",
    "    .to_numpy()\n",
    "  )\n",
    "\n",
    "  # compute correlation\n",
    "  corr = between_cancer_corr(\n",
    "    exparray,\n",
    "    mutationarray,\n",
    "    shuffle_graph,\n",
    "    cancertype_filter,\n",
    "    progressbar=False,\n",
    "  )\n",
    "  # save results\n",
    "  corr.to_feather(shuffled_dir+f\"shuffle{i}.feather\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Shuffle OLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutationarrays = dict(\n",
    "   mb = (\n",
    "    mutationtab\n",
    "    .apply(ols, axis=0, args=(mutation_burden,))\n",
    "    .fillna(0)\n",
    "    .to_numpy()\n",
    "  ),\n",
    "   mb_specific = (\n",
    "    mutationtab\n",
    "    .apply(ols, axis=0, args=(specific_mutation_burden, True))\n",
    "    .fillna(0)\n",
    "    .to_numpy()\n",
    "  ),\n",
    ")\n",
    "print(\"mutationarray mb:\", mutationarrays[\"mb\"].shape)\n",
    "print(\"mutationarray mb_specific:\", mutationarrays[\"mb_specific\"].shape)\n",
    "print(\"shuffle graph:\", shuffle_graph.shape)\n",
    "print(\"cancertype_filter:\", cancertype_filter.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(secrets.randbits(128))\n",
    "check_dir(outputdata+\"shuffle_bct_mb_ols/\")\n",
    "check_dir(outputdata+\"shuffle_bct_mb_specific_ols/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 234739320750563714933441023972028553133\n",
    "rng = np.random.default_rng(seed)\n",
    "spawn = rng.spawn(100)\n",
    "index = exptab.index.to_numpy()\n",
    "\n",
    "for i, child_rng in enumerate(tqdm(spawn)):\n",
    "  # shuffle expression data\n",
    "  shuffled_index = pd.Index(child_rng.permutation(index), name=\"cancer_type\")\n",
    "  exparray = (\n",
    "    exptab\n",
    "    .set_index(shuffled_index)\n",
    "    .groupby([\"cancer_type\"])\n",
    "    .mean()\n",
    "    .sort_index()\n",
    "    .to_numpy()\n",
    "  )\n",
    "  for norm, mutationarray in mutationarrays.items():\n",
    "    # compute correlation\n",
    "    corr = between_cancer_corr(\n",
    "      exparray,\n",
    "      mutationarray,\n",
    "      shuffle_graph,\n",
    "      cancertype_filter,\n",
    "      progressbar=False,\n",
    "    )\n",
    "    # save results\n",
    "    corr.to_feather(outputdata+f\"shuffle_bct_{norm}_ols/shuffle{i+1}.feather\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Cancer-specific driver BCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncg_drivers = pd.read_table(\n",
    "  rawdata+\"NCG_cancerdrivers_annotation_supporting_evidence.tsv\"\n",
    ")\n",
    "print(ncg_drivers.shape)\n",
    "ncg_drivers.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Export data for cancer type matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncg_cancertypes = (\n",
    "  ncg_drivers\n",
    "  .cancer_type\n",
    "  .dropna()\n",
    "  .drop_duplicates()\n",
    "  .reset_index(drop=True)\n",
    ")\n",
    "print(\"# cancer types:\", len(ncg_cancertypes))\n",
    "ncg_cancertypes.to_excel(outputdata+\"ncg_cancertypes.ods\", index=False)\n",
    "ncg_cancertypes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exptab = (\n",
    "  pd.read_feather(outputdata+\"expression.feather\")\n",
    "  .groupby([\"cancer_type\"])\n",
    "  .mean()\n",
    "  .sort_index()\n",
    ")\n",
    "print(exptab.shape)\n",
    "exptab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "match = (\n",
    "  pd.read_excel(outputdata+\"cancer_type_matching.ods\", sheet_name=\"NCG_mapping\")\n",
    "  .dropna()\n",
    "  .set_index(\"NCG_cancer_type\")\n",
    "  .merge(\n",
    "    ncg_drivers[[\"symbol\", \"cancer_type\"]],\n",
    "    left_index=True,\n",
    "    right_on=\"cancer_type\",\n",
    "    how=\"inner\",\n",
    "  )\n",
    "  .drop_duplicates()\n",
    ")\n",
    "match.to_csv(outputdata+\"cancer_type_matching.csv\", index=False)\n",
    "print(\"# NCG cancer types:\", match.cancer_type.nunique())\n",
    "print(\"# TCGA cancer types:\", match.TCGA_cancer_type.nunique())\n",
    "print(\"# NCG drivers:\", match.symbol.nunique())\n",
    "\n",
    "match = (\n",
    "  match\n",
    "  .assign(value=1)\n",
    "  .pivot_table(\n",
    "    index=\"TCGA_cancer_type\",\n",
    "    columns=\"symbol\",\n",
    "    values=\"value\",\n",
    "    fill_value=0,\n",
    "  )\n",
    "  # remove cancer types not in exptab\n",
    "  .loc[exptab.index, :]\n",
    "  # keep only drivers with 3 or more mutations\n",
    "  #.loc[:, lambda x: x.sum() > 2]\n",
    "  .astype(bool)\n",
    ")\n",
    "print(match.shape)\n",
    "match.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maingraph = pd.read_csv(outputdata+\"main_graph_filtered.csv\")\n",
    "drivers = np.intersect1d(maingraph.driver.unique(), match.columns)\n",
    "\n",
    "# filter main graph to only include drivers in NCG\n",
    "maingraph = maingraph[maingraph.driver.isin(drivers)]\n",
    "drivers = maingraph.driver.unique()\n",
    "neighbours = maingraph.neighbour.unique()\n",
    "print(\"# shuffle pairs:\", len(maingraph))\n",
    "print(\"# drivers:\", len(drivers))\n",
    "print(\"# neighbours:\", len(neighbours))\n",
    "maingraph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# refilter match and exptab\n",
    "exptab = exptab[neighbours]\n",
    "match = match[drivers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutationtab = pd.read_feather(outputdata+\"mutation.feather\")\n",
    "\n",
    "cancer_freq = (\n",
    "  mutationtab.index.to_frame(index=False)\n",
    "  .groupby(\"cancer_type\")\n",
    "  .size()\n",
    ")\n",
    "\n",
    "mutationtab = (\n",
    "  mutationtab[drivers]\n",
    "  .groupby([\"cancer_type\"])\n",
    "  .sum()\n",
    "  .div(cancer_freq, axis=0)\n",
    "  # order axis\n",
    "  .T.sort_index()\n",
    "  .T.sort_index()\n",
    ")\n",
    "print(cancer_freq.shape)\n",
    "print(mutationtab.shape)\n",
    "mutationtab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_burden = (\n",
    "  pd.read_csv(\n",
    "    outputdata+\"driver_specific_mutation_burden.csv\", index_col=\"cancer_type\")\n",
    "  [drivers]\n",
    ")\n",
    "print(mutation_burden.shape)\n",
    "mutation_burden.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"match:\", match.shape)\n",
    "print(\"exptab:\", exptab.shape)\n",
    "print(\"mutationtab:\", mutationtab.shape)\n",
    "print(\"mutation_burden:\", mutation_burden.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter to keep only cancer types where the driver has mutations\n",
    "# and is annotated in NCG as a cancer driver in the given cancer types\n",
    "cancertype_filter = (\n",
    "  mutationtab.T\n",
    "  .reset_index(names=\"driver\")\n",
    "  .set_index(\"driver\")\n",
    "  # remove cancer types with no mutations\n",
    "  .transform(lambda x: x > 0)\n",
    ")\n",
    "# keep only cancer types where the driver is annotated in NCG\n",
    "cancertype_filter = (\n",
    "  (cancertype_filter & match.T)\n",
    "  # remove drivers with no mutations in more than 2 cancer types\n",
    "  .loc[lambda x: x.sum(axis=1) > 2]\n",
    ")\n",
    "# drivers that passed all filters\n",
    "drivers = cancertype_filter.index.to_numpy()\n",
    "print(cancertype_filter.shape)\n",
    "nct = cancertype_filter.sum(axis=1)\n",
    "cancertype_filter.to_csv(outputdata+\"cancertype_filter_ncg.csv\", index=True)\n",
    "cancertype_filter.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# refilter dataframes\n",
    "maingraph = maingraph[maingraph.driver.isin(drivers)]\n",
    "# neighbours that passed all filters\n",
    "neighbours = maingraph.neighbour.unique()\n",
    "# create arrays\n",
    "exparray = exptab[neighbours].to_numpy()\n",
    "mutationarray = (mutationtab[drivers]/mutation_burden[drivers]).fillna(0).to_numpy()\n",
    "cancertype_filter = cancertype_filter.to_numpy()\n",
    "\n",
    "graph = (\n",
    "  maingraph\n",
    "  .assign(value=1)\n",
    "  .pivot_table(index=\"neighbour\", columns=\"driver\", values=\"value\", fill_value=0)\n",
    "  .astype(bool)\n",
    ")\n",
    "graph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"exparray:\", exparray.shape)\n",
    "print(\"mutationarray:\", mutationarray.shape)\n",
    "print(\"graph:\", graph.shape)\n",
    "print(\"cancertype_filter:\", cancertype_filter.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Compute Correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute correlation\n",
    "corr = between_cancer_corr(\n",
    "  exparray,\n",
    "  mutationarray,\n",
    "  graph,\n",
    "  cancertype_filter,\n",
    "  progressbar=True,\n",
    ").merge(nct.rename(\"nct\"), left_on=\"driver\", right_index=True)\n",
    "# save results\n",
    "corr.to_csv(outputdata+\"bct_corr_ncg_drivers.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## BCT correlation with mutation frequency groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>rho</th>\n",
       "      <th>rho_pval</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>nct</th>\n",
       "      <th>nmut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>-0.011765</td>\n",
       "      <td>0.965508</td>\n",
       "      <td>APOB</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.282353</td>\n",
       "      <td>0.289350</td>\n",
       "      <td>APOBEC1</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver       rho  rho_pval neighbour  nct  nmut\n",
       "0   A1CF -0.011765  0.965508      APOB   16    39\n",
       "1   A1CF  0.282353  0.289350   APOBEC1   16    39"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>rho_low_mut</th>\n",
       "      <th>rho_pval_low_mut</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>rho_high_mut</th>\n",
       "      <th>rho_pval_high_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.404762</td>\n",
       "      <td>0.319889</td>\n",
       "      <td>APOB</td>\n",
       "      <td>0.452381</td>\n",
       "      <td>0.260405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>-0.047619</td>\n",
       "      <td>0.910849</td>\n",
       "      <td>APOBEC1</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.910849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver  rho_low_mut  rho_pval_low_mut neighbour  rho_high_mut  \\\n",
       "0   A1CF     0.404762          0.319889      APOB      0.452381   \n",
       "1   A1CF    -0.047619          0.910849   APOBEC1      0.047619   \n",
       "\n",
       "   rho_pval_high_mut  \n",
       "0           0.260405  \n",
       "1           0.910849  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filted_bct_corr = pd.read_csv(outputdata+\"bct_corr_filtered.csv\")\n",
    "grouped_bct_corr = pd.read_csv(outputdata+\"bct_corr_mutation_groups.csv\")\n",
    "display(filted_bct_corr.head(2))\n",
    "grouped_bct_corr.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.merge(\n",
    "  filted_bct_corr,\n",
    "  grouped_bct_corr,\n",
    "  on=[\"driver\", \"neighbour\"],\n",
    ")[[\n",
    "  \"driver\", \"neighbour\", \"rho\", \"rho_low_mut\",\n",
    "  \"rho_high_mut\", \"rho_pval\", \"rho_pval_low_mut\", \"rho_pval_high_mut\"\n",
    "  ]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# pairs: 319863\n",
      "# significant pairs: 39421\n",
      "# significant pairs (low mutation): 23848\n",
      "# significant pairs (high mutation): 21378\n",
      "# significant pairs in common: 452\n",
      "\n",
      "All pairs\n",
      "\n",
      "Significant pairs\n",
      "\n",
      "Non-significant pairs\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>rho</th>\n",
       "      <th>pval</th>\n",
       "      <th>sign_concordance_pct</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pair_type</th>\n",
       "      <th>comparison</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">All pairs</th>\n",
       "      <th>rho vs rho_low_mut</th>\n",
       "      <td>0.455</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho vs rho_high_mut</th>\n",
       "      <td>0.401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>63.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho_low_mut vs rho_high_mut</th>\n",
       "      <td>-0.030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Significant pairs</th>\n",
       "      <th>rho vs rho_low_mut</th>\n",
       "      <td>0.677</td>\n",
       "      <td>0.0</td>\n",
       "      <td>85.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho vs rho_high_mut</th>\n",
       "      <td>0.652</td>\n",
       "      <td>0.0</td>\n",
       "      <td>82.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho_low_mut vs rho_high_mut</th>\n",
       "      <td>0.410</td>\n",
       "      <td>0.0</td>\n",
       "      <td>70.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Non-significant pairs</th>\n",
       "      <th>rho vs rho_low_mut</th>\n",
       "      <td>0.377</td>\n",
       "      <td>0.0</td>\n",
       "      <td>62.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho vs rho_high_mut</th>\n",
       "      <td>0.324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rho_low_mut vs rho_high_mut</th>\n",
       "      <td>-0.117</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     rho  pval  \\\n",
       "pair_type             comparison                                 \n",
       "All pairs             rho vs rho_low_mut           0.455   0.0   \n",
       "                      rho vs rho_high_mut          0.401   0.0   \n",
       "                      rho_low_mut vs rho_high_mut -0.030   0.0   \n",
       "Significant pairs     rho vs rho_low_mut           0.677   0.0   \n",
       "                      rho vs rho_high_mut          0.652   0.0   \n",
       "                      rho_low_mut vs rho_high_mut  0.410   0.0   \n",
       "Non-significant pairs rho vs rho_low_mut           0.377   0.0   \n",
       "                      rho vs rho_high_mut          0.324   0.0   \n",
       "                      rho_low_mut vs rho_high_mut -0.117   0.0   \n",
       "\n",
       "                                                   sign_concordance_pct  \n",
       "pair_type             comparison                                         \n",
       "All pairs             rho vs rho_low_mut                           65.5  \n",
       "                      rho vs rho_high_mut                          63.3  \n",
       "                      rho_low_mut vs rho_high_mut                  47.5  \n",
       "Significant pairs     rho vs rho_low_mut                           85.2  \n",
       "                      rho vs rho_high_mut                          82.3  \n",
       "                      rho_low_mut vs rho_high_mut                  70.6  \n",
       "Non-significant pairs rho vs rho_low_mut                           62.7  \n",
       "                      rho vs rho_high_mut                          60.6  \n",
       "                      rho_low_mut vs rho_high_mut                  44.3  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"# pairs:\", len(results))\n",
    "print(\"# significant pairs:\", (results.rho_pval < 0.05).sum())\n",
    "print(\"# significant pairs (low mutation):\", (results.rho_pval_low_mut < 0.05).sum())\n",
    "print(\"# significant pairs (high mutation):\", (results.rho_pval_high_mut < 0.05).sum())\n",
    "print(\"# significant pairs in common:\",\n",
    "      len(results[\n",
    "          (results.rho_pval < 0.05) &\n",
    "          (results.rho_pval_low_mut < 0.05) &\n",
    "          (results.rho_pval_high_mut < 0.05)\n",
    "          ]))\n",
    "stats = []\n",
    "for pair_type in [\"All pairs\", \"Significant pairs\", \"Non-significant pairs\"]:\n",
    "  print()\n",
    "  print(pair_type)\n",
    "  if pair_type == \"All pairs\":\n",
    "    data = results\n",
    "  elif pair_type == \"Significant pairs\":\n",
    "    data = results[results.rho_pval < 0.05]\n",
    "  else:\n",
    "    data = results[results.rho_pval >= 0.05]\n",
    "\n",
    "  concordance = data.assign(\n",
    "    rho=lambda x: x.rho/x.rho.abs(),\n",
    "    rho_low_mut=lambda x: x.rho_low_mut/x.rho_low_mut.abs(),\n",
    "    rho_high_mut=lambda x: x.rho_high_mut/x.rho_high_mut.abs(),\n",
    "  )\n",
    "\n",
    "  for comparison in [\n",
    "    (\"rho\", \"rho_low_mut\"),\n",
    "    (\"rho\", \"rho_high_mut\"),\n",
    "    (\"rho_low_mut\", \"rho_high_mut\"),\n",
    "  ]:\n",
    "    rho, pval = spstats.spearmanr(\n",
    "      data[comparison[0]], data[comparison[1]], alternative=\"two-sided\")\n",
    "    signconc = (\n",
    "      (concordance[comparison[0]] == concordance[comparison[1]]).sum()\n",
    "      /len(concordance)* 100\n",
    "    ).round(1)\n",
    "    stats.append({\n",
    "      \"pair_type\": pair_type,\n",
    "      \"comparison\": f\"{comparison[0]} vs {comparison[1]}\",\n",
    "      \"rho\": round(rho, 3),\n",
    "      \"pval\": round(pval, 3),\n",
    "      \"sign_concordance_pct\": signconc,\n",
    "    })\n",
    "pd.DataFrame(stats).set_index([\"pair_type\", \"comparison\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Cancer-specific driver BCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# NCG pairs: 47399\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>rho</th>\n",
       "      <th>rho_ngc</th>\n",
       "      <th>rho_pval</th>\n",
       "      <th>rho_pval_ngc</th>\n",
       "      <th>nct</th>\n",
       "      <th>nct_ngc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>ABCA2</td>\n",
       "      <td>-0.542727</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.006139</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>ABCB6</td>\n",
       "      <td>-0.018265</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.932493</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>ABCC8</td>\n",
       "      <td>-0.554468</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>ABCD4</td>\n",
       "      <td>-0.301370</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.152398</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>ABCG5</td>\n",
       "      <td>0.028267</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.895686</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver neighbour       rho  rho_ngc  rho_pval  rho_pval_ngc  nct  nct_ngc\n",
       "0  ABCC9     ABCA2 -0.542727     -0.5  0.006139      0.666667   24        3\n",
       "1  ABCC9     ABCB6 -0.018265      0.5  0.932493      0.666667   24        3\n",
       "2  ABCC9     ABCC8 -0.554468      1.0  0.004928      0.000000   24        3\n",
       "3  ABCC9     ABCD4 -0.301370      0.5  0.152398      0.666667   24        3\n",
       "4  ABCC9     ABCG5  0.028267      0.5  0.895686      0.666667   24        3"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"# NCG pairs:\", len(pd.read_csv(outputdata+\"bct_corr_ncg_drivers.csv\")))\n",
    "results = (\n",
    "  pd.read_csv(outputdata+\"bct_corr_filtered.csv\")\n",
    "  .merge(\n",
    "    pd.read_csv(outputdata+\"bct_corr_ncg_drivers.csv\"),\n",
    "    on=[\"driver\", \"neighbour\"],\n",
    "    suffixes=(\"\", \"_ngc\"),\n",
    "  )\n",
    "  [[\n",
    "    \"driver\", \"neighbour\", \"rho\", \"rho_ngc\", \"rho_pval\",\n",
    "    \"rho_pval_ngc\", \"nct\", \"nct_ngc\"\n",
    "  ]]\n",
    ")\n",
    "sign_intersection = results[\n",
    "  (results.rho_pval < 0.05) & (results.rho_pval_ngc < 0.05)\n",
    "]\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# pairs in common:\", len(results))\n",
    "print(\"# significant pairs (normal):\", (results.rho_pval < 0.05).sum())\n",
    "print(\"# significant pairs (NCG):\", (results.rho_pval_ngc < 0.05).sum())\n",
    "print(\"# significant pairs in common:\", len(sign_intersection))\n",
    "print(\"significant jaccard:\", len(sign_intersection) / (\n",
    "  (results.rho_pval < 0.05).sum() + (results.rho_pval_ngc < 0.05).sum()\n",
    "))\n",
    "rho, pval = spstats.spearmanr(\n",
    "  results.rho, results.rho_ngc, alternative=\"two-sided\"\n",
    ")\n",
    "print(f\"Spearman correlation: rho={rho}, p-value={pval}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.nct_ngc.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctmatching = pd.read_csv(outputdata+\"cancer_type_matching.csv\")\n",
    "ctmatching.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctstats = (\n",
    "  ctmatching\n",
    "  .groupby(\"symbol\", as_index=False)\n",
    "  .nunique()\n",
    "  .rename(\n",
    "    columns={\"cancer_type\": \"NCG_cancer_type\"}\n",
    "  )\n",
    "  [lambda x: x.symbol.isin(results.driver.unique())]\n",
    "  .reset_index(drop=True)\n",
    ")\n",
    "display(ctstats.head(2))\n",
    "ctstats.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = results[results.driver.isin(\n",
    "  ctstats.loc[ctstats.NCG_cancer_type > 7, \"symbol\"].unique()\n",
    ")]\n",
    "print(len(data))\n",
    "print(data.driver.nunique())\n",
    "spstats.spearmanr(\n",
    "  data.rho, data.rho_ngc, alternative=\"two-sided\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = results[[\"driver\", \"nct\", \"nct_ngc\"]].drop_duplicates().set_index(\"driver\")\n",
    "display(x.describe())\n",
    "(x.nct/x.nct_ngc).rename(\"ratio\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctfilter_ncg = pd.read_csv(outputdata+\"cancertype_filter_ncg.csv\", index_col=\"driver\")\n",
    "ctfilter = (\n",
    "  pd.read_csv(outputdata+\"cancertype_filter.csv\", index_col=\"driver\")\n",
    "  .loc[ctfilter_ncg.index]\n",
    ")\n",
    "mutationtab = (\n",
    "  pd.read_feather(outputdata+\"mutation.feather\")\n",
    "  [ctfilter.index]\n",
    ")\n",
    "mutationtab = (\n",
    "  mutationtab\n",
    "  .groupby([\"cancer_type\"])\n",
    "  .sum()\n",
    "  # order axis\n",
    "  .T.sort_index()\n",
    "  .sort_index(axis=1)\n",
    ")\n",
    "print(ctfilter_ncg.shape, ctfilter.shape, mutationtab.shape)\n",
    "display((\n",
    " mutationtab[ctfilter_ncg].mean(axis=1)/mutationtab[ctfilter].mean(axis=1)\n",
    ").describe())\n",
    "spstats.mannwhitneyu(\n",
    "  mutationtab[ctfilter_ncg].mean(axis=1), \n",
    "  mutationtab[ctfilter].mean(axis=1),\n",
    "  alternative=\"greater\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## BCT Pathway Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "324638\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>fake_neighbour</th>\n",
       "      <th>jaccard</th>\n",
       "      <th>mean_pagerank_fake</th>\n",
       "      <th>median_pagerank_fake</th>\n",
       "      <th>mean_pagerank_real</th>\n",
       "      <th>median_pagerank_real</th>\n",
       "      <th>distance</th>\n",
       "      <th>reg_TFs</th>\n",
       "      <th>n_shortest_paths</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>TOR1AIP1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SPI1,TP53</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>ATP2B1</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CEBPA,E2F4,EGR1,ESR1,GATA6,HNF4A,PPARA,PPARG,R...</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>5.0</td>\n",
       "      <td>E2F4,ETS1,MITF,MYC</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>LLGL2</td>\n",
       "      <td>PTK6</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ESR1,SOX9</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>SNX27</td>\n",
       "      <td>SLC5A11</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CEBPA,HNF4A</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver neighbour fake_neighbour   jaccard  mean_pagerank_fake  \\\n",
       "0  ABCA7     ADRB2       TOR1AIP1  1.000000            0.000748   \n",
       "1  ABCA7     APOA1         ATP2B1  0.375000            0.000113   \n",
       "2  ABCA7    HNRNPD           E2F8  0.600000            0.000100   \n",
       "3  ABCA7     LLGL2           PTK6  0.545455            0.000172   \n",
       "4  ABCA7     SNX27        SLC5A11  0.666667            0.000073   \n",
       "\n",
       "   median_pagerank_fake  mean_pagerank_real  median_pagerank_real  distance  \\\n",
       "0              0.000748            0.000748              0.000748       5.0   \n",
       "1              0.000058            0.000111              0.000066       5.0   \n",
       "2              0.000147            0.000278              0.000149       5.0   \n",
       "3              0.000049            0.000089              0.000042       5.0   \n",
       "4              0.000073            0.000064              0.000063       5.0   \n",
       "\n",
       "                                             reg_TFs  n_shortest_paths  \n",
       "0                                          SPI1,TP53               2.0  \n",
       "1  CEBPA,E2F4,EGR1,ESR1,GATA6,HNF4A,PPARA,PPARG,R...              10.0  \n",
       "2                                 E2F4,ETS1,MITF,MYC               4.0  \n",
       "3                                          ESR1,SOX9               2.0  \n",
       "4                                        CEBPA,HNF4A               2.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_analysis = pd.read_csv(outputdata+\"full_pathway_analysis.csv\")\n",
    "print(len(pathway_analysis))\n",
    "pathway_analysis.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>rho</th>\n",
       "      <th>rho_pval</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>nct</th>\n",
       "      <th>nmut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>-0.011765</td>\n",
       "      <td>0.965508</td>\n",
       "      <td>APOB</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.282353</td>\n",
       "      <td>0.289350</td>\n",
       "      <td>APOBEC1</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.297059</td>\n",
       "      <td>0.263863</td>\n",
       "      <td>APOBEC2</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.270588</td>\n",
       "      <td>0.310761</td>\n",
       "      <td>APOBEC3A</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.247059</td>\n",
       "      <td>0.356275</td>\n",
       "      <td>APOBEC3B</td>\n",
       "      <td>16</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver       rho  rho_pval neighbour  nct  nmut\n",
       "0   A1CF -0.011765  0.965508      APOB   16    39\n",
       "1   A1CF  0.282353  0.289350   APOBEC1   16    39\n",
       "2   A1CF  0.297059  0.263863   APOBEC2   16    39\n",
       "3   A1CF  0.270588  0.310761  APOBEC3A   16    39\n",
       "4   A1CF  0.247059  0.356275  APOBEC3B   16    39"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "associations = pd.read_csv(outputdata+\"bct_corr_mb_specific.csv\")\n",
    "associations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>rho</th>\n",
       "      <th>rho_pval</th>\n",
       "      <th>neighbour</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>0.127820</td>\n",
       "      <td>0.591246</td>\n",
       "      <td>ATP2B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>-0.153383</td>\n",
       "      <td>0.518524</td>\n",
       "      <td>CSNK1G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>-0.236090</td>\n",
       "      <td>0.316294</td>\n",
       "      <td>E2F8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>-0.093233</td>\n",
       "      <td>0.695829</td>\n",
       "      <td>MEAF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>-0.359398</td>\n",
       "      <td>0.119633</td>\n",
       "      <td>MKNK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver       rho  rho_pval neighbour\n",
       "0  ABCA7  0.127820  0.591246    ATP2B1\n",
       "1  ABCA7 -0.153383  0.518524   CSNK1G2\n",
       "2  ABCA7 -0.236090  0.316294      E2F8\n",
       "3  ABCA7 -0.093233  0.695829     MEAF6\n",
       "4  ABCA7 -0.359398  0.119633     MKNK1"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fake_associations = pd.read_csv(outputdata+\"bct_corr_fake_pairs.csv\")\n",
    "fake_associations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3283\n"
     ]
    }
   ],
   "source": [
    "tf_neighbour_pairs = pathway_analysis.loc[\n",
    "  lambda x: x.distance == 1, [\"driver\", \"neighbour\", \"fake_neighbour\"]\n",
    "]\n",
    "print(len(tf_neighbour_pairs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# associations: 455441\n",
      "# associations (p-value < 0.05): 54460\n",
      "proportion of significant associations: 0.11957641055592272\n",
      "proportion of significant associations with TFs: 0.15016752969844654\n",
      "# fake associations: 262578\n",
      "# fake associations (p-value < 0.05): 30204\n",
      "proportion of significant fake associations: 0.11502867719306263\n"
     ]
    }
   ],
   "source": [
    "print(\"# associations:\", len(associations))\n",
    "print(\"# associations (p-value < 0.05):\",\n",
    "      len(associations[lambda x: x[\"rho_pval\"] < 0.05]))\n",
    "print(\"proportion of significant associations:\",\n",
    "      len(associations[lambda x: x[\"rho_pval\"] < 0.05]) / len(associations))\n",
    "print(\"proportion of significant associations with TFs:\",\n",
    "      len(associations[lambda x: x[\"rho_pval\"] < 0.05]\n",
    "          .merge(tf_neighbour_pairs, on=[\"driver\", \"neighbour\"])) /\n",
    "      len(tf_neighbour_pairs))\n",
    "print(\"# fake associations:\", len(fake_associations))\n",
    "print(\"# fake associations (p-value < 0.05):\",\n",
    "      len(fake_associations[lambda x: x[\"rho_pval\"] < 0.05]))\n",
    "print(\"proportion of significant fake associations:\",\n",
    "      len(fake_associations[lambda x: x[\"rho_pval\"] < 0.05]) / len(fake_associations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "324638\n",
      "324565.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>sign</th>\n",
       "      <th>pagerank_type</th>\n",
       "      <th>pagerank_value</th>\n",
       "      <th>pair_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>real</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>real</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>real</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>LLGL2</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>real</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>SNX27</td>\n",
       "      <td>True</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>real</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver neighbour   sign pagerank_type  pagerank_value pair_type\n",
       "0  ABCA7     ADRB2  False          mean        0.000748      real\n",
       "1  ABCA7     APOA1  False          mean        0.000111      real\n",
       "2  ABCA7    HNRNPD  False          mean        0.000278      real\n",
       "3  ABCA7     LLGL2  False          mean        0.000089      real\n",
       "4  ABCA7     SNX27   True          mean        0.000064      real"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_associations = (\n",
    "  associations\n",
    "  .merge(pathway_analysis, on=[\"driver\", \"neighbour\"], how=\"inner\")\n",
    "  .assign(sign=lambda x: x[\"rho_pval\"] < 0.05)\n",
    "  [[\n",
    "    \"driver\", \"neighbour\", \"mean_pagerank_real\", \"median_pagerank_real\", \"sign\"\n",
    "  ]]\n",
    "  .melt(\n",
    "    id_vars=[\"driver\", \"neighbour\", \"sign\"],\n",
    "    var_name=\"pagerank_type\",\n",
    "    value_name=\"pagerank_value\"\n",
    "  )\n",
    "  .assign(\n",
    "    pagerank_type=lambda x: x[\"pagerank_type\"].str.split(\"_\").str[0],\n",
    "  )\n",
    "  .assign(pair_type=\"real\")\n",
    "  .drop_duplicates()\n",
    ")\n",
    "print(len(pathway_analysis))\n",
    "print(len(pathway_associations)/2)\n",
    "pathway_associations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "324638\n",
      "262578.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>driver</th>\n",
       "      <th>neighbour</th>\n",
       "      <th>sign</th>\n",
       "      <th>pagerank_type</th>\n",
       "      <th>pagerank_value</th>\n",
       "      <th>pair_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ATP2B1</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>CSNK1G2</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>MEAF6</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>MKNK1</td>\n",
       "      <td>False</td>\n",
       "      <td>mean</td>\n",
       "      <td>0.000501</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  driver neighbour   sign pagerank_type  pagerank_value pair_type\n",
       "0  ABCA7    ATP2B1  False          mean        0.000113      fake\n",
       "1  ABCA7   CSNK1G2  False          mean        0.000130      fake\n",
       "2  ABCA7      E2F8  False          mean        0.000100      fake\n",
       "3  ABCA7     MEAF6  False          mean        0.000049      fake\n",
       "4  ABCA7     MKNK1  False          mean        0.000501      fake"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fake_pathway_associations = (\n",
    "  fake_associations\n",
    "  .merge(\n",
    "    pathway_analysis,\n",
    "    left_on=[\"driver\", \"neighbour\"],\n",
    "    right_on=[\"driver\", \"fake_neighbour\"],\n",
    "    how=\"inner\",\n",
    "    suffixes=(\"_fake\", \"\")\n",
    "  )\n",
    "  .drop(columns=[\"neighbour_fake\", \"neighbour\"])\n",
    "  .assign(sign=lambda x: x[\"rho_pval\"] < 0.05)\n",
    "  [[\n",
    "    \"driver\", \"fake_neighbour\", \"mean_pagerank_fake\", \"median_pagerank_fake\", \"sign\"\n",
    "  ]]\n",
    "  .drop_duplicates()\n",
    "  .melt(\n",
    "    id_vars=[\"driver\", \"fake_neighbour\", \"sign\"],\n",
    "    var_name=\"pagerank_type\",\n",
    "    value_name=\"pagerank_value\"\n",
    "  )\n",
    "  .assign(\n",
    "    pagerank_type=lambda x: x[\"pagerank_type\"].str.split(\"_\").str[0],\n",
    "  )\n",
    "  .rename(columns={\"fake_neighbour\": \"neighbour\"})\n",
    "  .assign(pair_type=\"fake\")\n",
    ")\n",
    "print(len(pathway_analysis))\n",
    "print(len(fake_pathway_associations)/2)\n",
    "fake_pathway_associations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_associations = (\n",
    "  pd.concat([pathway_associations, fake_pathway_associations], ignore_index=True)\n",
    ")\n",
    "print(len(all_associations))\n",
    "all_associations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "quantiles = (\n",
    "  all_associations\n",
    "  .groupby([\"pair_type\", \"pagerank_type\"])\n",
    "  [\"pagerank_value\"]\n",
    "  .quantile(np.arange(0.1, 1.1, 0.1))\n",
    "  .reset_index(name=\"quantile_value\")\n",
    "  .rename(columns={\"level_2\": \"quantile\"})\n",
    ")\n",
    "quantiles[lambda x: x[\"quantile\"]==0.5].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groups = (\n",
    "  all_associations\n",
    "  .merge(quantiles, on=[\"pair_type\", \"pagerank_type\"])\n",
    "  [lambda x: x[\"pagerank_value\"] <= x[\"quantile_value\"]]\n",
    "  .groupby(\n",
    "    [\n",
    "      \"driver\", \"neighbour\", \"pair_type\", \"pagerank_type\", \"pagerank_value\",\n",
    "      \"sign\"\n",
    "    ],\n",
    "    as_index=False\n",
    "  )\n",
    "  [\"quantile\"].min()\n",
    ")\n",
    "print(all_associations.shape[0])\n",
    "groups.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sign = (\n",
    "  groups[groups[\"sign\"]]\n",
    "  .groupby([\"pair_type\", \"pagerank_type\", \"quantile\"], as_index=False)\n",
    "  .size()\n",
    "  .pivot(\n",
    "    index=[\"pair_type\", \"pagerank_type\",],\n",
    "    columns=\"quantile\",\n",
    "    values=\"size\"\n",
    "  )\n",
    ")\n",
    "all = (\n",
    "  groups\n",
    "  .groupby([\"pair_type\", \"pagerank_type\", \"quantile\"], as_index=False)\n",
    "  .size()\n",
    "  .pivot(\n",
    "    index=[\"pair_type\", \"pagerank_type\",],\n",
    "    columns=\"quantile\",\n",
    "    values=\"size\"\n",
    "  )\n",
    ")\n",
    "(sign/all).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean = (\n",
    "  pd.qcut(\n",
    "    (\n",
    "      pathway_associations\n",
    "      [lambda x: x.pagerank_type==\"mean\"]\n",
    "      .drop(columns=[\"pagerank_type\", \"pair_type\"])\n",
    "      .set_index([\"driver\", \"neighbour\"])\n",
    "      [\"pagerank_value\"]\n",
    "    ),\n",
    "    q=10, duplicates=\"drop\"\n",
    "  )\n",
    "  .reset_index(name=\"quantile\")\n",
    "  .merge(pathway_associations[[\"driver\", \"neighbour\", \"sign\"]].drop_duplicates(), on=[\"driver\", \"neighbour\"])\n",
    "  .groupby([\"quantile\", \"sign\"], as_index=False)\n",
    "  .size()\n",
    "  .set_index(\"quantile\")\n",
    ")\n",
    "mean.loc[mean.sign, \"size\"]/mean.groupby(\"quantile\")[\"size\"].sum()"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "qmd,ipynb"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
