{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import altair as alt\n",
    "from pyspark.sql import SparkSession\n",
    "alt.data_transformers.enable(\"vegafusion\")\n",
    "plt.rcParams.update({\"figure.dpi\": 120})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_dir(dir: str):\n",
    "    \"\"\"\n",
    "    Creates a given path driectory \"dir\" if it does not exist.\n",
    "    Args:\n",
    "        dir (str): Path to the directory. \n",
    "    \"\"\"\n",
    "    if os.path.exists(dir) and os.path.isdir(dir):\n",
    "        pass\n",
    "    else:\n",
    "        os.makedirs(dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "datadir = \"OpenTargets/\"\n",
    "check_dir(datadir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "suptabel1 = pd.read_csv(datadir + \"supptable1.csv\", sep=\",\")\n",
    "suptabel2 = pd.read_csv(datadir + \"supptable2.csv\", sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273\n",
      "43\n"
     ]
    }
   ],
   "source": [
    "print(len(suptabel1))\n",
    "print(len(suptabel2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "247\n",
      "39\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates based on Group.1\n",
    "suptabel1 = suptabel1.drop_duplicates(subset=[\"Group.1\"])\n",
    "print(len(suptabel1))\n",
    "\n",
    "suptabel2 = suptabel2.drop_duplicates(subset=[\"Group.1\"])\n",
    "print(len(suptabel2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "286\n"
     ]
    }
   ],
   "source": [
    "# Merge the two tables putting the first table first and the second table second \n",
    "supptable = pd.concat([suptabel1, suptabel2], axis=0)\n",
    "print(len(supptable))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  gene_id   x    n       bin        sd      prob           pbn         z  pos  \\\n",
      "0   ACTR3  18   66  3.633663  2.816817  0.055056  1.142413e-08  5.100203   18   \n",
      "1   ANXA2  20  109  6.009901  3.694577  0.055137  2.057292e-06  3.786658   20   \n",
      "2   ASF1B   9   22  1.316832  1.303309  0.059856  2.394384e-06  5.895125    9   \n",
      "3  ATAD3A  18   64  3.534653  2.528100  0.055229  7.014642e-09  5.721826   18   \n",
      "4  ATP1B3  10   27  1.376238  1.247810  0.050972  4.470500e-07  6.911118    9   \n",
      "\n",
      "   neg  ...          survsig     psurvsig         nsurvsig  survsign  \\\n",
      "0    0  ...      significant  significant  non significant  positive   \n",
      "1    0  ...      significant  significant  non significant  positive   \n",
      "2    0  ...      significant  significant  non significant  positive   \n",
      "3    0  ...      significant  significant  non significant  positive   \n",
      "4    1  ...  non significant  significant  non significant  positive   \n",
      "\n",
      "   survposfrac  kdriver  canonical  oncogene  tsg  type  \n",
      "0     0.875000        0          0         0    0   BCT  \n",
      "1     0.900000        1          0         0    0   BCT  \n",
      "2     0.727273        0          0         0    0   BCT  \n",
      "3     0.777778        0          0         0    0   BCT  \n",
      "4     1.000000        0          0         0    0   BCT  \n",
      "\n",
      "[5 rows x 32 columns]\n"
     ]
    }
   ],
   "source": [
    "# Change the name \"Group.1\" to \"gene_id\"\n",
    "supptable = supptable.rename(columns={\"Group.1\": \"gene_id\"})\n",
    "print(supptable.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Open Targets Search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "targetsPath = datadir+\"targets\"\n",
    "moleculePath = datadir+\"molecule\"\n",
    "associationPath = datadir+\"associationByOverallDirect\"\n",
    "diseasePath = datadir+\"diseases\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Establish spark connection\n",
    "spark = SparkSession.builder.master('local[*]').getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read datasets\n",
    "targets = spark.read.parquet(targetsPath)\n",
    "diseases = spark.read.parquet(diseasePath)\n",
    "molecules = spark.read.parquet(moleculePath)\n",
    "association = spark.read.parquet(associationPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- id: string (nullable = true)\n",
      " |-- approvedSymbol: string (nullable = true)\n",
      " |-- biotype: string (nullable = true)\n",
      " |-- transcriptIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- canonicalTranscript: struct (nullable = true)\n",
      " |    |-- id: string (nullable = true)\n",
      " |    |-- chromosome: string (nullable = true)\n",
      " |    |-- start: long (nullable = true)\n",
      " |    |-- end: long (nullable = true)\n",
      " |    |-- strand: string (nullable = true)\n",
      " |-- canonicalExons: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- genomicLocation: struct (nullable = true)\n",
      " |    |-- chromosome: string (nullable = true)\n",
      " |    |-- start: long (nullable = true)\n",
      " |    |-- end: long (nullable = true)\n",
      " |    |-- strand: integer (nullable = true)\n",
      " |-- alternativeGenes: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- approvedName: string (nullable = true)\n",
      " |-- go: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |    |    |-- evidence: string (nullable = true)\n",
      " |    |    |-- aspect: string (nullable = true)\n",
      " |    |    |-- geneProduct: string (nullable = true)\n",
      " |    |    |-- ecoId: string (nullable = true)\n",
      " |-- hallmarks: struct (nullable = true)\n",
      " |    |-- attributes: array (nullable = true)\n",
      " |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |-- pmid: long (nullable = true)\n",
      " |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |-- attribute_name: string (nullable = true)\n",
      " |    |-- cancerHallmarks: array (nullable = true)\n",
      " |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |-- pmid: long (nullable = true)\n",
      " |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |-- impact: string (nullable = true)\n",
      " |    |    |    |-- label: string (nullable = true)\n",
      " |-- synonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- symbolSynonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- nameSynonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- functionDescriptions: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- subcellularLocations: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- location: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |    |    |-- termSL: string (nullable = true)\n",
      " |    |    |-- labelSL: string (nullable = true)\n",
      " |-- targetClass: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: long (nullable = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- level: string (nullable = true)\n",
      " |-- obsoleteSymbols: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- obsoleteNames: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- constraint: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- constraintType: string (nullable = true)\n",
      " |    |    |-- score: float (nullable = true)\n",
      " |    |    |-- exp: float (nullable = true)\n",
      " |    |    |-- obs: integer (nullable = true)\n",
      " |    |    |-- oe: float (nullable = true)\n",
      " |    |    |-- oeLower: float (nullable = true)\n",
      " |    |    |-- oeUpper: float (nullable = true)\n",
      " |    |    |-- upperRank: integer (nullable = true)\n",
      " |    |    |-- upperBin: integer (nullable = true)\n",
      " |    |    |-- upperBin6: integer (nullable = true)\n",
      " |-- tep: struct (nullable = true)\n",
      " |    |-- targetFromSourceId: string (nullable = true)\n",
      " |    |-- description: string (nullable = true)\n",
      " |    |-- therapeuticArea: string (nullable = true)\n",
      " |    |-- url: string (nullable = true)\n",
      " |-- proteinIds: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- dbXrefs: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- chemicalProbes: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- control: string (nullable = true)\n",
      " |    |    |-- drugId: string (nullable = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- isHighQuality: boolean (nullable = true)\n",
      " |    |    |-- mechanismOfAction: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- origin: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- probeMinerScore: long (nullable = true)\n",
      " |    |    |-- probesDrugsScore: long (nullable = true)\n",
      " |    |    |-- scoreInCells: long (nullable = true)\n",
      " |    |    |-- scoreInOrganisms: long (nullable = true)\n",
      " |    |    |-- targetFromSourceId: string (nullable = true)\n",
      " |    |    |-- urls: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- niceName: string (nullable = true)\n",
      " |    |    |    |    |-- url: string (nullable = true)\n",
      " |-- homologues: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- speciesId: string (nullable = true)\n",
      " |    |    |-- speciesName: string (nullable = true)\n",
      " |    |    |-- homologyType: string (nullable = true)\n",
      " |    |    |-- targetGeneId: string (nullable = true)\n",
      " |    |    |-- isHighConfidence: string (nullable = true)\n",
      " |    |    |-- targetGeneSymbol: string (nullable = true)\n",
      " |    |    |-- queryPercentageIdentity: double (nullable = true)\n",
      " |    |    |-- targetPercentageIdentity: double (nullable = true)\n",
      " |    |    |-- priority: integer (nullable = true)\n",
      " |-- tractability: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- modality: string (nullable = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- value: boolean (nullable = true)\n",
      " |-- safetyLiabilities: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- event: string (nullable = true)\n",
      " |    |    |-- eventId: string (nullable = true)\n",
      " |    |    |-- effects: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- direction: string (nullable = true)\n",
      " |    |    |    |    |-- dosing: string (nullable = true)\n",
      " |    |    |-- biosamples: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- cellFormat: string (nullable = true)\n",
      " |    |    |    |    |-- cellLabel: string (nullable = true)\n",
      " |    |    |    |    |-- tissueId: string (nullable = true)\n",
      " |    |    |    |    |-- tissueLabel: string (nullable = true)\n",
      " |    |    |-- datasource: string (nullable = true)\n",
      " |    |    |-- literature: string (nullable = true)\n",
      " |    |    |-- url: string (nullable = true)\n",
      " |    |    |-- studies: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |    |-- name: string (nullable = true)\n",
      " |    |    |    |    |-- type: string (nullable = true)\n",
      " |-- pathways: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- pathwayId: string (nullable = true)\n",
      " |    |    |-- pathway: string (nullable = true)\n",
      " |    |    |-- topLevelTerm: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Browse the targets schema\n",
    "targets.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+\n",
      "|             id|approvedSymbol|\n",
      "+---------------+--------------+\n",
      "|ENSG00000002586|          CD99|\n",
      "|ENSG00000015479|         MATR3|\n",
      "|ENSG00000037280|          FLT4|\n",
      "|ENSG00000038427|          VCAN|\n",
      "|ENSG00000050730|         TNIP3|\n",
      "+---------------+--------------+\n",
      "only showing top 5 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Select fields of interest\n",
    "targetSelect = (\n",
    "    targets.select(\n",
    "        \"id\",\n",
    "        \"approvedSymbol\"\n",
    "    )\n",
    ")\n",
    "targetSelect.show(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "286\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>x</th>\n",
       "      <th>n</th>\n",
       "      <th>bin</th>\n",
       "      <th>sd</th>\n",
       "      <th>prob</th>\n",
       "      <th>pbn</th>\n",
       "      <th>z</th>\n",
       "      <th>pos</th>\n",
       "      <th>...</th>\n",
       "      <th>survsig</th>\n",
       "      <th>psurvsig</th>\n",
       "      <th>nsurvsig</th>\n",
       "      <th>survsign</th>\n",
       "      <th>survposfrac</th>\n",
       "      <th>kdriver</th>\n",
       "      <th>canonical</th>\n",
       "      <th>oncogene</th>\n",
       "      <th>tsg</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000109805</td>\n",
       "      <td>NCAPG</td>\n",
       "      <td>14</td>\n",
       "      <td>50</td>\n",
       "      <td>2.900990</td>\n",
       "      <td>2.184056</td>\n",
       "      <td>0.05802</td>\n",
       "      <td>6.252813e-07</td>\n",
       "      <td>5.081835</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.90</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000138160</td>\n",
       "      <td>KIF11</td>\n",
       "      <td>20</td>\n",
       "      <td>54</td>\n",
       "      <td>3.663366</td>\n",
       "      <td>3.383126</td>\n",
       "      <td>0.06784</td>\n",
       "      <td>1.423412e-10</td>\n",
       "      <td>4.828857</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id gene_id   x   n       bin        sd     prob           pbn  \\\n",
       "0  ENSG00000109805   NCAPG  14  50  2.900990  2.184056  0.05802  6.252813e-07   \n",
       "1  ENSG00000138160   KIF11  20  54  3.663366  3.383126  0.06784  1.423412e-10   \n",
       "\n",
       "          z  pos  ...      survsig     psurvsig         nsurvsig  survsign  \\\n",
       "0  5.081835   14  ...  significant  significant  non significant  positive   \n",
       "1  4.828857   19  ...  significant  significant  non significant  positive   \n",
       "\n",
       "  survposfrac  kdriver  canonical  oncogene  tsg  type  \n",
       "0        0.90        1          0         0    0   BCT  \n",
       "1        0.75        0          0         0    0   BCT  \n",
       "\n",
       "[2 rows x 33 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_df = (\n",
    "    targetSelect\n",
    "    .toPandas()\n",
    "    .rename(columns={\n",
    "        \"approvedSymbol\": \"gene_id\",\n",
    "        \"id\": \"ensembl_id\"\n",
    "    })\n",
    "    .merge(supptable, on=\"gene_id\")\n",
    ")\n",
    "print(len(target_df))\n",
    "target_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "# See if exist \"ensembl_id\" duplicates\n",
    "print((target_df.ensembl_id.duplicated()).sum()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- id: string (nullable = true)\n",
      " |-- code: string (nullable = true)\n",
      " |-- dbXRefs: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- description: string (nullable = true)\n",
      " |-- name: string (nullable = true)\n",
      " |-- directLocationIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- obsoleteTerms: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- parents: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- synonyms: struct (nullable = true)\n",
      " |    |-- hasBroadSynonym: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- hasExactSynonym: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- hasNarrowSynonym: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- hasRelatedSynonym: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |-- ancestors: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- descendants: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- children: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- therapeuticAreas: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- indirectLocationIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- ontology: struct (nullable = true)\n",
      " |    |-- isTherapeuticArea: boolean (nullable = true)\n",
      " |    |-- leaf: boolean (nullable = true)\n",
      " |    |-- sources: struct (nullable = true)\n",
      " |    |    |-- url: string (nullable = true)\n",
      " |    |    |-- name: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Browse the disease schema\n",
    "diseases.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------------+--------------------+\n",
      "|         id|                name|           ancestors|\n",
      "+-----------+--------------------+--------------------+\n",
      "|EFO_0000255|angioimmunoblasti...|[MONDO_0000430, O...|\n",
      "|EFO_0000508|    genetic disorder|      [OTAR_0000018]|\n",
      "|EFO_0001054|             leprosy|[EFO_0009387, OTA...|\n",
      "|EFO_0004287|ventricular fibri...|[EFO_0003777, EFO...|\n",
      "|EFO_0004302|anthropometric me...|       [EFO_0001444]|\n",
      "+-----------+--------------------+--------------------+\n",
      "only showing top 5 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Select fields of interest\n",
    "diseaseSelect = (\n",
    "    diseases.select(\n",
    "        \"id\",\n",
    "        \"name\",\n",
    "        \"ancestors\"\n",
    "    )\n",
    ")\n",
    "diseaseSelect.show(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[MONDO_0004992](https://platform.opentargets.org/disease/MONDO_0004992) is the main identifier for cancer. Since this *Disease Category* includes in its `descendants` all cancer types, we will use it to filter targets and drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1715\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>MONDO_0002759</td>\n",
       "      <td>bladder verrucous carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>MONDO_0003086</td>\n",
       "      <td>thymic mucoepidermoid carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       disease_id                     disease_name\n",
       "74  MONDO_0002759      bladder verrucous carcinoma\n",
       "77  MONDO_0003086  thymic mucoepidermoid carcinoma"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_df = (\n",
    "    diseaseSelect\n",
    "    .toPandas()\n",
    "    .explode(column='ancestors')\n",
    "    .loc[lambda x: x.ancestors == \"MONDO_0004992\", [\"id\", \"name\"]]\n",
    "    .rename(columns={\"id\": \"disease_id\", \"name\": \"disease_name\"})\n",
    ")\n",
    "print(len(disease_df))\n",
    "disease_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adds MONDO_0004992 as its own descendant, so we can filter out diseases that are descendants of MONDO_0004992\n",
    "disease_df = pd.concat([disease_df, diseaseSelect.toPandas()[[\"id\", \"name\"]].rename(columns={\"id\": \"disease_id\", \"name\": \"disease_name\"})[lambda x: x.disease_id==\"MONDO_0004992\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- diseaseId: string (nullable = true)\n",
      " |-- targetId: string (nullable = true)\n",
      " |-- score: double (nullable = true)\n",
      " |-- evidenceCount: long (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Browse the evidence schema\n",
    "association.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------------+\n",
      "|       targetId|  diseaseId|               score|\n",
      "+---------------+-----------+--------------------+\n",
      "|ENSG00000004399|EFO_0000580|0.005174117965585...|\n",
      "|ENSG00000012048|EFO_0000580|0.017054967553679457|\n",
      "|ENSG00000026025|EFO_0000580|0.001478319418738...|\n",
      "|ENSG00000026103|EFO_0000580|0.029566388374776147|\n",
      "|ENSG00000035862|EFO_0000580|0.007391597093694037|\n",
      "+---------------+-----------+--------------------+\n",
      "only showing top 5 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Select fields of interest\n",
    "associationSelect = (\n",
    "    association.select(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"score\",\n",
    "    )\n",
    ")\n",
    "associationSelect.show(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16509\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>EFO_0000580</td>\n",
       "      <td>medullary breast carcinoma</td>\n",
       "      <td>0.017055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MCM6</td>\n",
       "      <td>EFO_0000580</td>\n",
       "      <td>medullary breast carcinoma</td>\n",
       "      <td>0.003696</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id   disease_id                disease_name     score\n",
       "0   BRCA1  EFO_0000580  medullary breast carcinoma  0.017055\n",
       "1    MCM6  EFO_0000580  medullary breast carcinoma  0.003696"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "association_df = (\n",
    "    associationSelect\n",
    "    .toPandas()\n",
    "    .rename(columns={\n",
    "        \"targetId\": \"ensembl_id\",\n",
    "        \"diseaseId\": \"disease_id\"\n",
    "    })\n",
    "    .merge(disease_df)\n",
    "    .merge(target_df)\n",
    "    [[\"gene_id\", \"disease_id\", \"disease_name\", \"score\"]]\n",
    ")\n",
    "print(len(association_df))\n",
    "association_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "association_df.to_csv(datadir+\"target_disease_association.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- id: string (nullable = true)\n",
      " |-- canonicalSmiles: string (nullable = true)\n",
      " |-- inchiKey: string (nullable = true)\n",
      " |-- drugType: string (nullable = true)\n",
      " |-- blackBoxWarning: boolean (nullable = true)\n",
      " |-- name: string (nullable = true)\n",
      " |-- yearOfFirstApproval: long (nullable = true)\n",
      " |-- maximumClinicalTrialPhase: double (nullable = true)\n",
      " |-- parentId: string (nullable = true)\n",
      " |-- hasBeenWithdrawn: boolean (nullable = true)\n",
      " |-- isApproved: boolean (nullable = true)\n",
      " |-- tradeNames: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- synonyms: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- crossReferences: map (nullable = true)\n",
      " |    |-- key: string\n",
      " |    |-- value: array (valueContainsNull = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |-- childChemblIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- linkedDiseases: struct (nullable = true)\n",
      " |    |-- rows: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- count: integer (nullable = true)\n",
      " |-- linkedTargets: struct (nullable = true)\n",
      " |    |-- rows: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- count: integer (nullable = true)\n",
      " |-- description: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Browse the molecule schema\n",
    "molecules.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------+-------------+----------+--------------------+--------------------+-------------------------+--------------+\n",
      "|           id|         name|isApproved|       linkedTargets|      linkedDiseases|maximumClinicalTrialPhase|      drugType|\n",
      "+-------------+-------------+----------+--------------------+--------------------+-------------------------+--------------+\n",
      "|CHEMBL1086582|CHEMBL1086582|      NULL|                NULL|                NULL|                     NULL|Small molecule|\n",
      "|CHEMBL1173055|    RUCAPARIB|      true|{[ENSG00000143799...|{[EFO_0003060, MO...|                      4.0|Small molecule|\n",
      "+-------------+-------------+----------+--------------------+--------------------+-------------------------+--------------+\n",
      "only showing top 2 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Select fields of interest\n",
    "moleculeSelect = (\n",
    "    molecules.select(\n",
    "        \"id\",\n",
    "        \"name\",\n",
    "        \"isApproved\",\n",
    "        \"linkedTargets\",\n",
    "        \"linkedDiseases\",\n",
    "        'maximumClinicalTrialPhase',\n",
    "        'drugType'\n",
    "    )\n",
    ")\n",
    "moleculeSelect.show(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8919\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>isApproved</th>\n",
       "      <th>maximumClinicalTrialPhase</th>\n",
       "      <th>drugType</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>x</th>\n",
       "      <th>n</th>\n",
       "      <th>...</th>\n",
       "      <th>survsig</th>\n",
       "      <th>psurvsig</th>\n",
       "      <th>nsurvsig</th>\n",
       "      <th>survsign</th>\n",
       "      <th>survposfrac</th>\n",
       "      <th>kdriver</th>\n",
       "      <th>canonical</th>\n",
       "      <th>oncogene</th>\n",
       "      <th>tsg</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>56</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>MONDO_0007254</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>56</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>MONDO_0004992</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>56</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000117461</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>15</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000117461</td>\n",
       "      <td>MONDO_0007254</td>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>15</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            drug    drug_name isApproved  maximumClinicalTrialPhase  \\\n",
       "0  CHEMBL1234354  PF-04691502      False                        2.0   \n",
       "1  CHEMBL1234354  PF-04691502      False                        2.0   \n",
       "2  CHEMBL1234354  PF-04691502      False                        2.0   \n",
       "3  CHEMBL1234354  PF-04691502      False                        2.0   \n",
       "4  CHEMBL1234354  PF-04691502      False                        2.0   \n",
       "\n",
       "         drugType       ensembl_id     disease_id gene_id   x    n  ...  \\\n",
       "0  Small molecule  ENSG00000145675    EFO_0003869  PIK3R1  56  168  ...   \n",
       "1  Small molecule  ENSG00000145675  MONDO_0007254  PIK3R1  56  168  ...   \n",
       "2  Small molecule  ENSG00000145675  MONDO_0004992  PIK3R1  56  168  ...   \n",
       "3  Small molecule  ENSG00000117461    EFO_0003869  PIK3R3  15   75  ...   \n",
       "4  Small molecule  ENSG00000117461  MONDO_0007254  PIK3R3  15   75  ...   \n",
       "\n",
       "           survsig         psurvsig     nsurvsig  survsign  survposfrac  \\\n",
       "0  non significant  non significant  significant  negative         0.00   \n",
       "1  non significant  non significant  significant  negative         0.00   \n",
       "2  non significant  non significant  significant  negative         0.00   \n",
       "3      significant  non significant  significant  negative         0.25   \n",
       "4      significant  non significant  significant  negative         0.25   \n",
       "\n",
       "   kdriver  canonical  oncogene  tsg type  \n",
       "0        1          1         0    1  BCT  \n",
       "1        1          1         0    1  BCT  \n",
       "2        1          1         0    1  BCT  \n",
       "3        1          0         0    0  BCT  \n",
       "4        1          0         0    0  BCT  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule_df = (\n",
    "    moleculeSelect\n",
    "    .toPandas()\n",
    "    .dropna(subset=[\"linkedTargets\", \"linkedDiseases\"])\n",
    "    .assign(ensembl_id=lambda x: x.linkedTargets.apply(\n",
    "        lambda x: x.asDict()['rows']))\n",
    "    .assign(disease_id=lambda x: x.linkedDiseases.apply(\n",
    "        lambda x: x.asDict()['rows']))\n",
    "    .drop(columns=['linkedTargets', 'linkedDiseases'])\n",
    "    .explode('ensembl_id')\n",
    "    .explode('disease_id')\n",
    "    .rename(columns={\n",
    "        \"id\": \"drug\",\n",
    "        \"name\": \"drug_name\"\n",
    "    })\n",
    "    .merge(target_df)\n",
    ")\n",
    "print(len(molecule_df))\n",
    "molecule_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge molecule_df with disease_df on disease_id\n",
    "cancer_associated_drugs = (\n",
    "    molecule_df\n",
    "    .merge(disease_df, on=\"disease_id\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>isApproved</th>\n",
       "      <th>maximumClinicalTrialPhase</th>\n",
       "      <th>drugType</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>x</th>\n",
       "      <th>n</th>\n",
       "      <th>...</th>\n",
       "      <th>psurvsig</th>\n",
       "      <th>nsurvsig</th>\n",
       "      <th>survsign</th>\n",
       "      <th>survposfrac</th>\n",
       "      <th>kdriver</th>\n",
       "      <th>canonical</th>\n",
       "      <th>oncogene</th>\n",
       "      <th>tsg</th>\n",
       "      <th>type</th>\n",
       "      <th>disease_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>MONDO_0007254</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>56</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>BCT</td>\n",
       "      <td>breast cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>MONDO_0004992</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>56</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>BCT</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000117461</td>\n",
       "      <td>MONDO_0007254</td>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>15</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>breast cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1234354</td>\n",
       "      <td>PF-04691502</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000117461</td>\n",
       "      <td>MONDO_0004992</td>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>15</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>non significant</td>\n",
       "      <td>significant</td>\n",
       "      <td>negative</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1683544</td>\n",
       "      <td>ERIBULIN MESYLATE</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>ENSG00000196230</td>\n",
       "      <td>EFO_0003060</td>\n",
       "      <td>TUBB</td>\n",
       "      <td>25</td>\n",
       "      <td>155</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>non-small cell lung carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4867</th>\n",
       "      <td>CHEMBL451887</td>\n",
       "      <td>CARFILZOMIB</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Protein</td>\n",
       "      <td>ENSG00000108671</td>\n",
       "      <td>MONDO_0008315</td>\n",
       "      <td>PSMD11</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>prostate cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4868</th>\n",
       "      <td>CHEMBL451887</td>\n",
       "      <td>CARFILZOMIB</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Protein</td>\n",
       "      <td>ENSG00000108671</td>\n",
       "      <td>MONDO_0002367</td>\n",
       "      <td>PSMD11</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>kidney cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4869</th>\n",
       "      <td>CHEMBL451887</td>\n",
       "      <td>CARFILZOMIB</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Protein</td>\n",
       "      <td>ENSG00000108671</td>\n",
       "      <td>EFO_0000702</td>\n",
       "      <td>PSMD11</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>small cell lung carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4870</th>\n",
       "      <td>CHEMBL451887</td>\n",
       "      <td>CARFILZOMIB</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Protein</td>\n",
       "      <td>ENSG00000108671</td>\n",
       "      <td>EFO_0000222</td>\n",
       "      <td>PSMD11</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4871</th>\n",
       "      <td>CHEMBL451887</td>\n",
       "      <td>CARFILZOMIB</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Protein</td>\n",
       "      <td>ENSG00000108671</td>\n",
       "      <td>EFO_1001051</td>\n",
       "      <td>PSMD11</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>significant</td>\n",
       "      <td>non significant</td>\n",
       "      <td>positive</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BCT</td>\n",
       "      <td>mycosis fungoides</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4872 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               drug          drug_name isApproved  maximumClinicalTrialPhase  \\\n",
       "0     CHEMBL1234354        PF-04691502      False                        2.0   \n",
       "1     CHEMBL1234354        PF-04691502      False                        2.0   \n",
       "2     CHEMBL1234354        PF-04691502      False                        2.0   \n",
       "3     CHEMBL1234354        PF-04691502      False                        2.0   \n",
       "4     CHEMBL1683544  ERIBULIN MESYLATE       True                        4.0   \n",
       "...             ...                ...        ...                        ...   \n",
       "4867   CHEMBL451887        CARFILZOMIB       True                        4.0   \n",
       "4868   CHEMBL451887        CARFILZOMIB       True                        4.0   \n",
       "4869   CHEMBL451887        CARFILZOMIB       True                        4.0   \n",
       "4870   CHEMBL451887        CARFILZOMIB       True                        4.0   \n",
       "4871   CHEMBL451887        CARFILZOMIB       True                        4.0   \n",
       "\n",
       "            drugType       ensembl_id     disease_id gene_id   x    n  ...  \\\n",
       "0     Small molecule  ENSG00000145675  MONDO_0007254  PIK3R1  56  168  ...   \n",
       "1     Small molecule  ENSG00000145675  MONDO_0004992  PIK3R1  56  168  ...   \n",
       "2     Small molecule  ENSG00000117461  MONDO_0007254  PIK3R3  15   75  ...   \n",
       "3     Small molecule  ENSG00000117461  MONDO_0004992  PIK3R3  15   75  ...   \n",
       "4     Small molecule  ENSG00000196230    EFO_0003060    TUBB  25  155  ...   \n",
       "...              ...              ...            ...     ...  ..  ...  ...   \n",
       "4867         Protein  ENSG00000108671  MONDO_0008315  PSMD11  19   58  ...   \n",
       "4868         Protein  ENSG00000108671  MONDO_0002367  PSMD11  19   58  ...   \n",
       "4869         Protein  ENSG00000108671    EFO_0000702  PSMD11  19   58  ...   \n",
       "4870         Protein  ENSG00000108671    EFO_0000222  PSMD11  19   58  ...   \n",
       "4871         Protein  ENSG00000108671    EFO_1001051  PSMD11  19   58  ...   \n",
       "\n",
       "             psurvsig         nsurvsig  survsign  survposfrac  kdriver  \\\n",
       "0     non significant      significant  negative     0.000000        1   \n",
       "1     non significant      significant  negative     0.000000        1   \n",
       "2     non significant      significant  negative     0.250000        1   \n",
       "3     non significant      significant  negative     0.250000        1   \n",
       "4         significant  non significant  positive     0.666667        0   \n",
       "...               ...              ...       ...          ...      ...   \n",
       "4867      significant  non significant  positive     0.900000        0   \n",
       "4868      significant  non significant  positive     0.900000        0   \n",
       "4869      significant  non significant  positive     0.900000        0   \n",
       "4870      significant  non significant  positive     0.900000        0   \n",
       "4871      significant  non significant  positive     0.900000        0   \n",
       "\n",
       "      canonical  oncogene  tsg  type                   disease_name  \n",
       "0             1         0    1   BCT                  breast cancer  \n",
       "1             1         0    1   BCT                         cancer  \n",
       "2             0         0    0   BCT                  breast cancer  \n",
       "3             0         0    0   BCT                         cancer  \n",
       "4             0         0    0   BCT  non-small cell lung carcinoma  \n",
       "...         ...       ...  ...   ...                            ...  \n",
       "4867          0         0    0   BCT                prostate cancer  \n",
       "4868          0         0    0   BCT                  kidney cancer  \n",
       "4869          0         0    0   BCT      small cell lung carcinoma  \n",
       "4870          0         0    0   BCT         acute myeloid leukemia  \n",
       "4871          0         0    0   BCT              mycosis fungoides  \n",
       "\n",
       "[4872 rows x 40 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_associated_drugs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final Considerations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PIK3R1' 'PIK3R3' 'TUBB' 'TUBA1B' 'TUBA1C' 'CHEK1' 'TOP2A' 'TYMS' 'CDK4'\n",
      " 'EZH2' 'PSMD14' 'PSMA1' 'PSMA7' 'PSMC2' 'PSMC4' 'PSMD11' 'CDK1' 'CDK2'\n",
      " 'AURKB' 'AURKA' 'RRM2' 'POLD1' 'POLE' 'POLA2' 'P4HB' 'KIF11' 'E2F1'\n",
      " 'E2F2' 'CRBN' 'ATP1B3' 'PLK1' 'VAMP2' 'TFRC' 'TTK' 'GMPS' 'PLK4' 'COL6A3'\n",
      " 'LAMB3']\n",
      "38\n"
     ]
    }
   ],
   "source": [
    "print(molecule_df.gene_id.unique())\n",
    "print(len(molecule_df.gene_id.unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We found that 38 of our selected neighbours (34 neighbours enriched in positive driver and survival associations and 4 enriched in negative driver and survival associations, supplementary tables 1 and 2) are known drug targets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "TUBB      678\n",
       "TUBA1B    672\n",
       "TUBA1C    672\n",
       "TOP2A     441\n",
       "TYMS      289\n",
       "RRM2      250\n",
       "POLD1     246\n",
       "POLA2     246\n",
       "POLE      246\n",
       "CDK4      162\n",
       "CRBN       84\n",
       "PSMA1      78\n",
       "PSMA7      78\n",
       "PSMD14     74\n",
       "PSMC2      74\n",
       "PSMC4      74\n",
       "PSMD11     74\n",
       "PIK3R1     71\n",
       "PIK3R3     71\n",
       "CDK2       64\n",
       "CHEK1      49\n",
       "AURKA      42\n",
       "AURKB      33\n",
       "CDK1       29\n",
       "KIF11      23\n",
       "PLK1       17\n",
       "EZH2       16\n",
       "ATP1B3      7\n",
       "PLK4        6\n",
       "TFRC        4\n",
       "TTK         2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_associated_drugs.gene_id.value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "31 were targeted by drugs developed for cancer treatment, confirming in this way the therapeutic potential of neighbours enriched in driver associations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "COL6A3    21\n",
       "P4HB      17\n",
       "VAMP2      9\n",
       "LAMB3      8\n",
       "GMPS       3\n",
       "E2F1       1\n",
       "E2F2       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Removing from molecule_df those that are in cancer_associated_drugs see how many are left with unique gene_ids \n",
    "molecule_df[~molecule_df.gene_id.isin(cancer_associated_drugs.gene_id)].gene_id.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "COL6A3    2\n",
       "E2F1      1\n",
       "E2F2      1\n",
       "GMPS      1\n",
       "LAMB3     1\n",
       "P4HB      4\n",
       "VAMP2     1\n",
       "Name: drug, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Proteins without drugs with an association with cancer\n",
    "# Drugs that can be used for drug repurposing\n",
    "drug_repurposing = (\n",
    "    molecule_df\n",
    "    .groupby(\"gene_id\")\n",
    "    # Add cancer and neoplasm ids and descendant diseases to exclude drugs with those associations\n",
    "    .filter(lambda x: x.disease_id.isin(\n",
    "        disease_df.disease_id.tolist() + [\"EFO_0000616\"]).sum() == 0)\n",
    "    .groupby(\"gene_id\")[\"drug\"]\n",
    "    .nunique()\n",
    ")\n",
    "drug_repurposing.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interestingly, 7 neighbours (_COL6A3_, _E2F1_, _E2F2_, _GMPS_, _LAMB3_, _PH4B_ and _VAMP2_) were targeted by drugs for non-cancer diseases, suggesting an opportunity for drug repurposing.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "molecule_df.to_csv(datadir+\"molecule_target_disease_association.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tese",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
